// openclaw/dist/plugin-sdk/memory-core-host-runtime-core.js
import { _ as resolveStateDir } from "../paths-yyDPxM31.js";
import { w as parseAgentSessionKey } from "../session-key-BR3Z-ljs.js";
import { m as resolveDefaultAgentId, v as resolveSessionAgentId } from "../agent-scope-CXWTwwic.js";
import { a as loadConfig } from "../io-CS2J_l4V.js";
import { o as parseNonNegativeByteSize } from "../zod-schema-C3jh3SvI.js";
import { c as resolveSessionTranscriptsDirForAgent } from "../paths-UazeViO92.js";
import { c as jsonResult, d as readNumberParam, h as readStringParam } from "../common-B7pbdYUb.js";
import { n as SILENT_REPLY_TOKEN } from "../tokens-wOGzQgw2.js";
import { t as DEFAULT_PI_COMPACTION_RESERVE_TOKENS_FLOOR } from "../pi-settings-DF7KuKBq.js";
import { t as resolveMemorySearchConfig } from "../memory-search-DIQ9kV2j.js";
import { n as resolveCronStyleNow } from "../current-time-C1zXvUcB.js";
import { n as emptyPluginConfigSchema } from "../config-schema-BTVf9GZX.js";
import "../memory-core-host-runtime-core-CP2NzfCG.js";
export { DEFAULT_PI_COMPACTION_RESERVE_TOKENS_FLOOR, SILENT_REPLY_TOKEN, emptyPluginConfigSchema, jsonResult, loadConfig, parseAgentSessionKey, parseNonNegativeByteSize, readNumberParam, readStringParam, resolveCronStyleNow, resolveDefaultAgentId, resolveMemorySearchConfig, resolveSessionAgentId, resolveSessionTranscriptsDirForAgent, resolveStateDir };
