// @line/bot-sdk/dist/exceptions.js
export class SignatureValidationFailed extends Error {
    signature;
    constructor(message, { signature } = {}) {
        super(message);
        this.name = this.constructor.name;
        Object.assign(this, { signature });
    }
}
export class JSONParseError extends Error {
    raw;
    constructor(message, { raw } = {}) {
        super(message);
        this.name = this.constructor.name;
        Object.assign(this, { raw });
    }
}
/**
 * @deprecated Legacy axios-based HTTP client only. `LineBotClient` (fetch-based) does not
 * throw this error. Network-level failures in `LineBotClient` surface as a native `TypeError`
 * from the Fetch API, not as `RequestError`.
 */
export class RequestError extends Error {
    code;
    originalError;
    constructor(message, { code, originalError }) {
        super(message);
        this.name = this.constructor.name;
        Object.assign(this, { code, originalError });
    }
}
/**
 * @deprecated Legacy axios-based HTTP client only. `LineBotClient` (fetch-based) does not
 * throw this error. Read failures in `LineBotClient` propagate unwrapped from the Fetch API.
 */
export class ReadError extends Error {
    originalError;
    constructor(message, { originalError }) {
        super(message);
        this.name = this.constructor.name;
        Object.assign(this, { originalError });
    }
}
/**
 * @deprecated Use `HTTPFetchError` instead. Thrown by the legacy axios-based HTTP client;
 * `LineBotClient` uses the Fetch API and throws `HTTPFetchError` on non-2xx responses.
 * Key differences: `HTTPFetchError` exposes `status`/`statusText`/`headers`/`body`
 * instead of `statusCode`/`statusMessage`/`originalError`.
 */
export class HTTPError extends Error {
    statusCode;
    statusMessage;
    originalError;
    constructor(message, { statusCode, statusMessage, originalError }) {
        super(message);
        this.name = this.constructor.name;
        Object.assign(this, { statusCode, statusMessage, originalError });
    }
}
export class HTTPFetchError extends Error {
    status;
    statusText;
    headers;
    body;
    constructor(message, { status, statusText, headers, body }) {
        super(message);
        this.name = this.constructor.name;
        Object.assign(this, { status, statusText, headers, body });
    }
}
