// @line/bot-sdk/dist/cjs/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OAuth = void 0;
const http_axios_js_1 = require("./http-axios.js");
const Types = require("./types.js");
const utils_js_1 = require("./utils.js");
const endpoints_js_1 = require("./endpoints.js");
/**
 * @deprecated Use clients generated by openapi spec instead.
 */
class Client {
    constructor(config) {
        this.requestOption = {};
        if (!config.channelAccessToken) {
            throw new Error("no channel access token");
        }
        this.config = config;
        this.http = new http_axios_js_1.default(Object.assign({ defaultHeaders: {
                Authorization: "Bearer " + this.config.channelAccessToken,
            }, responseParser: this.parseHTTPResponse.bind(this) }, config.httpConfig));
    }
    /**
     * @deprecated No direct equivalent in LineBotClient. Pass `xLineRetryKey` as the 2nd argument to `pushMessage`, `multicast`, `narrowcast`, or `broadcast` directly.
     */
    setRequestOptionOnce(option) {
        this.requestOption = option;
    }
    generateRequestConfig() {
        const config = { headers: {} };
        if (this.requestOption.retryKey) {
            config.headers["X-Line-Retry-Key"] = this.requestOption.retryKey;
        }
        // clear requestOption
        this.requestOption = {};
        return config;
    }
    parseHTTPResponse(response) {
        const { LINE_REQUEST_ID_HTTP_HEADER_NAME } = Types;
        let resBody = Object.assign({}, response.data);
        if (response.headers[LINE_REQUEST_ID_HTTP_HEADER_NAME]) {
            resBody[LINE_REQUEST_ID_HTTP_HEADER_NAME] =
                response.headers[LINE_REQUEST_ID_HTTP_HEADER_NAME];
        }
        return resBody;
    }
    /**
     * @deprecated Use {@link LineBotClient.pushMessage} instead.
     * @example
     * // Before:
     * client.pushMessage(to, messages, notificationDisabled, customAggregationUnits);
     * // After:
     * lineBotClient.pushMessage({ to, messages, notificationDisabled, customAggregationUnits }, xLineRetryKey);
     */
    pushMessage(to, messages, notificationDisabled = false, customAggregationUnits) {
        return this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/message/push`, {
            messages: (0, utils_js_1.toArray)(messages),
            to,
            notificationDisabled,
            customAggregationUnits,
        }, this.generateRequestConfig());
    }
    /**
     * @deprecated Use {@link LineBotClient.replyMessage} instead.
     * @example
     * // Before:
     * client.replyMessage(replyToken, messages, notificationDisabled);
     * // After:
     * lineBotClient.replyMessage({ replyToken, messages, notificationDisabled });
     */
    replyMessage(replyToken, messages, notificationDisabled = false) {
        return this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/message/reply`, {
            messages: (0, utils_js_1.toArray)(messages),
            replyToken,
            notificationDisabled,
        });
    }
    /**
     * @deprecated Use {@link LineBotClient.multicast} instead.
     * @example
     * // Before:
     * client.multicast(to, messages, notificationDisabled, customAggregationUnits);
     * // After:
     * lineBotClient.multicast({ to, messages, notificationDisabled, customAggregationUnits }, xLineRetryKey);
     */
    async multicast(to, messages, notificationDisabled = false, customAggregationUnits) {
        return this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/message/multicast`, {
            messages: (0, utils_js_1.toArray)(messages),
            to,
            notificationDisabled,
            customAggregationUnits,
        }, this.generateRequestConfig());
    }
    /**
     * @deprecated Use {@link LineBotClient.narrowcast} instead.
     * @example
     * // Before:
     * client.narrowcast(messages, recipient, filter, limit, notificationDisabled);
     * // After:
     * lineBotClient.narrowcast({ messages, recipient, filter, limit, notificationDisabled }, xLineRetryKey);
     */
    async narrowcast(messages, recipient, filter, limit, notificationDisabled) {
        return this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/message/narrowcast`, {
            messages: (0, utils_js_1.toArray)(messages),
            recipient,
            filter,
            limit,
            notificationDisabled,
        }, this.generateRequestConfig());
    }
    /**
     * @deprecated Use {@link LineBotClient.broadcast} instead.
     * @example
     * // Before:
     * client.broadcast(messages, notificationDisabled);
     * // After:
     * lineBotClient.broadcast({ messages, notificationDisabled }, xLineRetryKey);
     */
    async broadcast(messages, notificationDisabled = false) {
        return this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/message/broadcast`, {
            messages: (0, utils_js_1.toArray)(messages),
            notificationDisabled,
        }, this.generateRequestConfig());
    }
    /**
     * @deprecated Use {@link LineBotClient.validatePush} instead.
     * @example
     * // Before:
     * client.validatePushMessageObjects(messages);
     * // After:
     * lineBotClient.validatePush({ messages });
     */
    validatePushMessageObjects(messages) {
        return this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/message/validate/push`, {
            messages: (0, utils_js_1.toArray)(messages),
        }, this.generateRequestConfig());
    }
    /**
     * @deprecated Use {@link LineBotClient.validateReply} instead.
     * @example
     * // Before:
     * client.validateReplyMessageObjects(messages);
     * // After:
     * lineBotClient.validateReply({ messages });
     */
    validateReplyMessageObjects(messages) {
        return this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/message/validate/reply`, {
            messages: (0, utils_js_1.toArray)(messages),
        });
    }
    /**
     * @deprecated Use {@link LineBotClient.validateMulticast} instead.
     * @example
     * // Before:
     * client.validateMulticastMessageObjects(messages);
     * // After:
     * lineBotClient.validateMulticast({ messages });
     */
    async validateMulticastMessageObjects(messages) {
        return this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/message/validate/multicast`, {
            messages: (0, utils_js_1.toArray)(messages),
        }, this.generateRequestConfig());
    }
    /**
     * @deprecated Use {@link LineBotClient.validateNarrowcast} instead.
     * @example
     * // Before:
     * client.validateNarrowcastMessageObjects(messages);
     * // After:
     * lineBotClient.validateNarrowcast({ messages });
     */
    async validateNarrowcastMessageObjects(messages) {
        return this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/message/validate/narrowcast`, {
            messages: (0, utils_js_1.toArray)(messages),
        }, this.generateRequestConfig());
    }
    /**
     * @deprecated Use {@link LineBotClient.validateBroadcast} instead.
     * @example
     * // Before:
     * client.validateBroadcastMessageObjects(messages);
     * // After:
     * lineBotClient.validateBroadcast({ messages });
     */
    async validateBroadcastMessageObjects(messages) {
        return this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/message/validate/broadcast`, {
            messages: (0, utils_js_1.toArray)(messages),
        }, this.generateRequestConfig());
    }
    /**
     * @deprecated No direct equivalent in LineBotClient. See docs/guide/migration.md for details.
     */
    validateCustomAggregationUnits(units) {
        const messages = [];
        if (units.length > 1) {
            messages.push("customAggregationUnits can only contain one unit");
        }
        units.forEach((unit, index) => {
            if (unit.length > 30) {
                messages.push(`customAggregationUnits[${index}] must be less than or equal to 30 characters`);
            }
            if (!/^[a-zA-Z0-9_]+$/.test(unit)) {
                messages.push(`customAggregationUnits[${index}] must be alphanumeric characters or underscores`);
            }
        });
        return {
            messages,
            valid: messages.length === 0,
        };
    }
    /**
     * @deprecated Use {@link LineBotClient.getProfile} instead.
     * @example
     * // Before:
     * client.getProfile(userId);
     * // After:
     * lineBotClient.getProfile(userId);
     */
    async getProfile(userId) {
        const profile = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/profile/${userId}`);
        return (0, utils_js_1.ensureJSON)(profile);
    }
    async getChatMemberProfile(chatType, chatId, userId) {
        const profile = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/${chatType}/${chatId}/member/${userId}`);
        return (0, utils_js_1.ensureJSON)(profile);
    }
    /**
     * @deprecated Use {@link LineBotClient.getGroupMemberProfile} instead.
     * @example
     * // Before:
     * client.getGroupMemberProfile(groupId, userId);
     * // After:
     * lineBotClient.getGroupMemberProfile(groupId, userId);
     */
    async getGroupMemberProfile(groupId, userId) {
        return this.getChatMemberProfile("group", groupId, userId);
    }
    /**
     * @deprecated Use {@link LineBotClient.getRoomMemberProfile} instead.
     * @example
     * // Before:
     * client.getRoomMemberProfile(roomId, userId);
     * // After:
     * lineBotClient.getRoomMemberProfile(roomId, userId);
     */
    async getRoomMemberProfile(roomId, userId) {
        return this.getChatMemberProfile("room", roomId, userId);
    }
    async getChatMemberIds(chatType, chatId) {
        let memberIds = [];
        let start;
        do {
            const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/${chatType}/${chatId}/members/ids`, start ? { start } : null);
            (0, utils_js_1.ensureJSON)(res);
            memberIds = memberIds.concat(res.memberIds);
            start = res.next;
        } while (start);
        return memberIds;
    }
    /**
     * @deprecated Use {@link LineBotClient.getGroupMembersIds} instead. Note: the new method does not auto-paginate.
     * @example
     * // Before:
     * client.getGroupMemberIds(groupId);
     * // After:
     * lineBotClient.getGroupMembersIds(groupId);
     */
    async getGroupMemberIds(groupId) {
        return this.getChatMemberIds("group", groupId);
    }
    /**
     * @deprecated Use {@link LineBotClient.getRoomMembersIds} instead. Note: the new method does not auto-paginate.
     * @example
     * // Before:
     * client.getRoomMemberIds(roomId);
     * // After:
     * lineBotClient.getRoomMembersIds(roomId);
     */
    async getRoomMemberIds(roomId) {
        return this.getChatMemberIds("room", roomId);
    }
    /**
     * @deprecated Use {@link LineBotClient.getFollowers} instead. Note: the new method does not auto-paginate.
     * The old method collected all pages and returned a flat `string[]`.
     * The new method returns a single page (`GetFollowersResponse`); call it in a loop using the `next` token to retrieve all followers.
     * @example
     * // Before:
     * const allIds = await client.getBotFollowersIds();
     * // After (manual pagination):
     * let next: string | undefined;
     * const allIds: string[] = [];
     * do {
     *   const res = await lineBotClient.getFollowers(next);
     *   allIds.push(...res.userIds);
     *   next = res.next;
     * } while (next);
     */
    async getBotFollowersIds() {
        let userIds = [];
        let start;
        do {
            const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/followers/ids`, start ? { start, limit: 1000 } : { limit: 1000 });
            (0, utils_js_1.ensureJSON)(res);
            userIds = userIds.concat(res.userIds);
            start = res.next;
        } while (start);
        return userIds;
    }
    /**
     * @deprecated Use {@link LineBotClient.getGroupMemberCount} instead.
     * @example
     * // Before:
     * client.getGroupMembersCount(groupId);
     * // After:
     * lineBotClient.getGroupMemberCount(groupId);
     */
    async getGroupMembersCount(groupId) {
        const groupMemberCount = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/group/${groupId}/members/count`);
        return (0, utils_js_1.ensureJSON)(groupMemberCount);
    }
    /**
     * @deprecated Use {@link LineBotClient.getRoomMemberCount} instead.
     * @example
     * // Before:
     * client.getRoomMembersCount(roomId);
     * // After:
     * lineBotClient.getRoomMemberCount(roomId);
     */
    async getRoomMembersCount(roomId) {
        const roomMemberCount = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/room/${roomId}/members/count`);
        return (0, utils_js_1.ensureJSON)(roomMemberCount);
    }
    /**
     * @deprecated Use {@link LineBotClient.getGroupSummary} instead.
     * @example
     * // Before:
     * client.getGroupSummary(groupId);
     * // After:
     * lineBotClient.getGroupSummary(groupId);
     */
    async getGroupSummary(groupId) {
        const groupSummary = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/group/${groupId}/summary`);
        return (0, utils_js_1.ensureJSON)(groupSummary);
    }
    /**
     * @deprecated Use {@link LineBotClient.getMessageContent} instead.
     * @example
     * // Before:
     * client.getMessageContent(messageId);
     * // After:
     * lineBotClient.getMessageContent(messageId);
     */
    async getMessageContent(messageId) {
        return this.http.getStream(`${endpoints_js_1.DATA_API_PREFIX}/message/${messageId}/content`);
    }
    leaveChat(chatType, chatId) {
        return this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/${chatType}/${chatId}/leave`);
    }
    /**
     * @deprecated Use {@link LineBotClient.leaveGroup} instead.
     * @example
     * // Before:
     * client.leaveGroup(groupId);
     * // After:
     * lineBotClient.leaveGroup(groupId);
     */
    async leaveGroup(groupId) {
        return this.leaveChat("group", groupId);
    }
    /**
     * @deprecated Use {@link LineBotClient.leaveRoom} instead.
     * @example
     * // Before:
     * client.leaveRoom(roomId);
     * // After:
     * lineBotClient.leaveRoom(roomId);
     */
    async leaveRoom(roomId) {
        return this.leaveChat("room", roomId);
    }
    /**
     * @deprecated Use {@link LineBotClient.getRichMenu} instead.
     * @example
     * // Before:
     * client.getRichMenu(richMenuId);
     * // After:
     * lineBotClient.getRichMenu(richMenuId);
     */
    async getRichMenu(richMenuId) {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/richmenu/${richMenuId}`);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.createRichMenu} instead. Note: return type changed from `string` to `RichMenuIdResponse`.
     * @example
     * // Before:
     * client.createRichMenu(richMenu);
     * // After:
     * lineBotClient.createRichMenu(richMenu);
     */
    async createRichMenu(richMenu) {
        const res = await this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/richmenu`, richMenu);
        return (0, utils_js_1.ensureJSON)(res).richMenuId;
    }
    /**
     * @deprecated Use {@link LineBotClient.deleteRichMenu} instead.
     * @example
     * // Before:
     * client.deleteRichMenu(richMenuId);
     * // After:
     * lineBotClient.deleteRichMenu(richMenuId);
     */
    async deleteRichMenu(richMenuId) {
        return this.http.delete(`${endpoints_js_1.MESSAGING_API_PREFIX}/richmenu/${richMenuId}`);
    }
    /**
     * @deprecated Use {@link LineBotClient.getRichMenuAliasList} instead.
     * @example
     * // Before:
     * client.getRichMenuAliasList();
     * // After:
     * lineBotClient.getRichMenuAliasList();
     */
    async getRichMenuAliasList() {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/richmenu/alias/list`);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.getRichMenuAlias} instead.
     * @example
     * // Before:
     * client.getRichMenuAlias(richMenuAliasId);
     * // After:
     * lineBotClient.getRichMenuAlias(richMenuAliasId);
     */
    async getRichMenuAlias(richMenuAliasId) {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/richmenu/alias/${richMenuAliasId}`);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.createRichMenuAlias} instead.
     * @example
     * // Before:
     * client.createRichMenuAlias(richMenuId, richMenuAliasId);
     * // After:
     * lineBotClient.createRichMenuAlias({ richMenuId, richMenuAliasId });
     */
    async createRichMenuAlias(richMenuId, richMenuAliasId) {
        const res = await this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/richmenu/alias`, {
            richMenuId,
            richMenuAliasId,
        });
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.deleteRichMenuAlias} instead.
     * @example
     * // Before:
     * client.deleteRichMenuAlias(richMenuAliasId);
     * // After:
     * lineBotClient.deleteRichMenuAlias(richMenuAliasId);
     */
    async deleteRichMenuAlias(richMenuAliasId) {
        const res = this.http.delete(`${endpoints_js_1.MESSAGING_API_PREFIX}/richmenu/alias/${richMenuAliasId}`);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.updateRichMenuAlias} instead.
     * @example
     * // Before:
     * client.updateRichMenuAlias(richMenuAliasId, richMenuId);
     * // After:
     * lineBotClient.updateRichMenuAlias(richMenuAliasId, { richMenuId });
     */
    async updateRichMenuAlias(richMenuAliasId, richMenuId) {
        const res = await this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/richmenu/alias/${richMenuAliasId}`, {
            richMenuId,
        });
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.getRichMenuIdOfUser} instead. Note: return type changed from `string` to `RichMenuIdResponse`.
     * @example
     * // Before:
     * client.getRichMenuIdOfUser(userId);
     * // After:
     * lineBotClient.getRichMenuIdOfUser(userId);
     */
    async getRichMenuIdOfUser(userId) {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/user/${userId}/richmenu`);
        return (0, utils_js_1.ensureJSON)(res).richMenuId;
    }
    /**
     * @deprecated Use {@link LineBotClient.linkRichMenuIdToUser} instead.
     * @example
     * // Before:
     * client.linkRichMenuToUser(userId, richMenuId);
     * // After:
     * lineBotClient.linkRichMenuIdToUser(userId, richMenuId);
     */
    async linkRichMenuToUser(userId, richMenuId) {
        return this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/user/${userId}/richmenu/${richMenuId}`);
    }
    /**
     * @deprecated Use {@link LineBotClient.unlinkRichMenuIdFromUser} instead.
     * @example
     * // Before:
     * client.unlinkRichMenuFromUser(userId);
     * // After:
     * lineBotClient.unlinkRichMenuIdFromUser(userId);
     */
    async unlinkRichMenuFromUser(userId) {
        return this.http.delete(`${endpoints_js_1.MESSAGING_API_PREFIX}/user/${userId}/richmenu`);
    }
    /**
     * @deprecated Use {@link LineBotClient.linkRichMenuIdToUsers} instead.
     * @example
     * // Before:
     * client.linkRichMenuToMultipleUsers(richMenuId, userIds);
     * // After:
     * lineBotClient.linkRichMenuIdToUsers({ richMenuId, userIds });
     */
    async linkRichMenuToMultipleUsers(richMenuId, userIds) {
        return this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/richmenu/bulk/link`, {
            richMenuId,
            userIds,
        });
    }
    /**
     * @deprecated Use {@link LineBotClient.unlinkRichMenuIdFromUsers} instead.
     * @example
     * // Before:
     * client.unlinkRichMenusFromMultipleUsers(userIds);
     * // After:
     * lineBotClient.unlinkRichMenuIdFromUsers({ userIds });
     */
    async unlinkRichMenusFromMultipleUsers(userIds) {
        return this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/richmenu/bulk/unlink`, {
            userIds,
        });
    }
    /**
     * @deprecated Use {@link LineBotClient.getRichMenuImage} instead.
     * @example
     * // Before:
     * client.getRichMenuImage(richMenuId);
     * // After:
     * lineBotClient.getRichMenuImage(richMenuId);
     */
    async getRichMenuImage(richMenuId) {
        return this.http.getStream(`${endpoints_js_1.DATA_API_PREFIX}/richmenu/${richMenuId}/content`);
    }
    /**
     * @deprecated Use {@link LineBotClient.setRichMenuImage} instead. Note: the new method accepts a Blob as the second argument.
     * @example
     * // Before:
     * client.setRichMenuImage(richMenuId, data, contentType);
     * // After:
     * lineBotClient.setRichMenuImage(richMenuId, body);
     */
    async setRichMenuImage(richMenuId, data, contentType) {
        return this.http.postBinary(`${endpoints_js_1.DATA_API_PREFIX}/richmenu/${richMenuId}/content`, data, contentType);
    }
    /**
     * @deprecated Use {@link LineBotClient.getRichMenuList} instead.
     * @example
     * // Before:
     * client.getRichMenuList();
     * // After:
     * lineBotClient.getRichMenuList();
     */
    async getRichMenuList() {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/richmenu/list`);
        return (0, utils_js_1.ensureJSON)(res).richmenus;
    }
    /**
     * @deprecated Use {@link LineBotClient.setDefaultRichMenu} instead.
     * @example
     * // Before:
     * client.setDefaultRichMenu(richMenuId);
     * // After:
     * lineBotClient.setDefaultRichMenu(richMenuId);
     */
    async setDefaultRichMenu(richMenuId) {
        return this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/user/all/richmenu/${richMenuId}`);
    }
    /**
     * @deprecated Use {@link LineBotClient.getDefaultRichMenuId} instead. Note: return type changed from `string` to `RichMenuIdResponse`.
     * @example
     * // Before:
     * client.getDefaultRichMenuId();
     * // After:
     * lineBotClient.getDefaultRichMenuId();
     */
    async getDefaultRichMenuId() {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/user/all/richmenu`);
        return (0, utils_js_1.ensureJSON)(res).richMenuId;
    }
    /**
     * @deprecated Use {@link LineBotClient.cancelDefaultRichMenu} instead.
     * @example
     * // Before:
     * client.deleteDefaultRichMenu();
     * // After:
     * lineBotClient.cancelDefaultRichMenu();
     */
    async deleteDefaultRichMenu() {
        return this.http.delete(`${endpoints_js_1.MESSAGING_API_PREFIX}/user/all/richmenu`);
    }
    /**
     * @deprecated Use {@link LineBotClient.issueLinkToken} instead. Note: return type changed from `string` to `IssueLinkTokenResponse`.
     * @example
     * // Before:
     * client.getLinkToken(userId);
     * // After:
     * lineBotClient.issueLinkToken(userId);
     */
    async getLinkToken(userId) {
        const res = await this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/user/${userId}/linkToken`);
        return (0, utils_js_1.ensureJSON)(res).linkToken;
    }
    /**
     * @deprecated Use {@link LineBotClient.getNumberOfSentReplyMessages} instead.
     * @example
     * // Before:
     * client.getNumberOfSentReplyMessages(date);
     * // After:
     * lineBotClient.getNumberOfSentReplyMessages(date);
     */
    async getNumberOfSentReplyMessages(date) {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/message/delivery/reply?date=${date}`);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.getNumberOfSentPushMessages} instead.
     * @example
     * // Before:
     * client.getNumberOfSentPushMessages(date);
     * // After:
     * lineBotClient.getNumberOfSentPushMessages(date);
     */
    async getNumberOfSentPushMessages(date) {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/message/delivery/push?date=${date}`);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.getNumberOfSentMulticastMessages} instead.
     * @example
     * // Before:
     * client.getNumberOfSentMulticastMessages(date);
     * // After:
     * lineBotClient.getNumberOfSentMulticastMessages(date);
     */
    async getNumberOfSentMulticastMessages(date) {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/message/delivery/multicast?date=${date}`);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.getNarrowcastProgress} instead.
     * @example
     * // Before:
     * client.getNarrowcastProgress(requestId);
     * // After:
     * lineBotClient.getNarrowcastProgress(requestId);
     */
    async getNarrowcastProgress(requestId) {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/message/progress/narrowcast?requestId=${requestId}`);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.getMessageQuota} instead.
     * @example
     * // Before:
     * client.getTargetLimitForAdditionalMessages();
     * // After:
     * lineBotClient.getMessageQuota();
     */
    async getTargetLimitForAdditionalMessages() {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/message/quota`);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.getMessageQuotaConsumption} instead.
     * @example
     * // Before:
     * client.getNumberOfMessagesSentThisMonth();
     * // After:
     * lineBotClient.getMessageQuotaConsumption();
     */
    async getNumberOfMessagesSentThisMonth() {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/message/quota/consumption`);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.getNumberOfSentBroadcastMessages} instead.
     * @example
     * // Before:
     * client.getNumberOfSentBroadcastMessages(date);
     * // After:
     * lineBotClient.getNumberOfSentBroadcastMessages(date);
     */
    async getNumberOfSentBroadcastMessages(date) {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/message/delivery/broadcast?date=${date}`);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.getNumberOfMessageDeliveries} instead.
     * @example
     * // Before:
     * client.getNumberOfMessageDeliveries(date);
     * // After:
     * lineBotClient.getNumberOfMessageDeliveries(date);
     */
    async getNumberOfMessageDeliveries(date) {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/insight/message/delivery?date=${date}`);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.getNumberOfFollowers} instead. Note: `date` is now optional.
     * @example
     * // Before:
     * client.getNumberOfFollowers(date);
     * // After:
     * lineBotClient.getNumberOfFollowers(date);
     */
    async getNumberOfFollowers(date) {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/insight/followers?date=${date}`);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.getFriendsDemographics} instead.
     * @example
     * // Before:
     * client.getFriendDemographics();
     * // After:
     * lineBotClient.getFriendsDemographics();
     */
    async getFriendDemographics() {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/insight/demographic`);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.getMessageEvent} instead.
     * @example
     * // Before:
     * client.getUserInteractionStatistics(requestId);
     * // After:
     * lineBotClient.getMessageEvent(requestId);
     */
    async getUserInteractionStatistics(requestId) {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/insight/message/event?requestId=${requestId}`);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.getStatisticsPerUnit} instead.
     * @example
     * // Before:
     * client.getStatisticsPerUnit(unit, from, to);
     * // After:
     * lineBotClient.getStatisticsPerUnit(unit, from, to);
     */
    async getStatisticsPerUnit(customAggregationUnit, from, to) {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/insight/message/event/aggregation?customAggregationUnit=${customAggregationUnit}&from=${from}&to=${to}`);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.createAudienceGroup} instead.
     * @example
     * // Before:
     * client.createUploadAudienceGroup({...});
     * // After:
     * lineBotClient.createAudienceGroup({...});
     */
    async createUploadAudienceGroup(uploadAudienceGroup) {
        const res = await this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/audienceGroup/upload`, Object.assign({}, uploadAudienceGroup));
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.createAudienceForUploadingUserIds} instead.
     * @example
     * // Before:
     * client.createUploadAudienceGroupByFile({ ..., file });
     * // After:
     * lineBotClient.createAudienceForUploadingUserIds(file, description, isIfaAudience, uploadDescription);
     */
    async createUploadAudienceGroupByFile(uploadAudienceGroup) {
        const file = await this.http.toBuffer(uploadAudienceGroup.file);
        const body = (0, utils_js_1.createMultipartFormData)(Object.assign(Object.assign({}, uploadAudienceGroup), { file }));
        const res = await this.http.postFormMultipart(`${endpoints_js_1.DATA_API_PREFIX}/audienceGroup/upload/byFile`, body);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.addAudienceToAudienceGroup} instead.
     * Note: if you were passing `description`, you must also call
     * {@link LineBotClient.updateAudienceGroupDescription} separately, as
     * `AddAudienceToAudienceGroupRequest` does not accept a `description` field.
     * @example
     * // Before:
     * client.updateUploadAudienceGroup({ audienceGroupId, description, uploadDescription, audiences });
     * // After (two calls required when description is used):
     * await lineBotClient.addAudienceToAudienceGroup({ audienceGroupId, uploadDescription, audiences });
     * await lineBotClient.updateAudienceGroupDescription(audienceGroupId, { description });
     */
    async updateUploadAudienceGroup(uploadAudienceGroup, 
    // for set request timeout
    httpConfig) {
        const res = await this.http.put(`${endpoints_js_1.MESSAGING_API_PREFIX}/audienceGroup/upload`, Object.assign({}, uploadAudienceGroup), httpConfig);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.addUserIdsToAudience} instead.
     * @example
     * // Before:
     * client.updateUploadAudienceGroupByFile({ ..., file });
     * // After:
     * lineBotClient.addUserIdsToAudience(file, audienceGroupId, uploadDescription);
     */
    async updateUploadAudienceGroupByFile(uploadAudienceGroup, 
    // for set request timeout
    httpConfig) {
        const file = await this.http.toBuffer(uploadAudienceGroup.file);
        const body = (0, utils_js_1.createMultipartFormData)(Object.assign(Object.assign({}, uploadAudienceGroup), { file }));
        const res = await this.http.putFormMultipart(`${endpoints_js_1.DATA_API_PREFIX}/audienceGroup/upload/byFile`, body, httpConfig);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.createClickBasedAudienceGroup} instead.
     * @example
     * // Before:
     * client.createClickAudienceGroup({...});
     * // After:
     * lineBotClient.createClickBasedAudienceGroup({...});
     */
    async createClickAudienceGroup(clickAudienceGroup) {
        const res = await this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/audienceGroup/click`, Object.assign({}, clickAudienceGroup));
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.createImpBasedAudienceGroup} instead.
     * @example
     * // Before:
     * client.createImpAudienceGroup({...});
     * // After:
     * lineBotClient.createImpBasedAudienceGroup({...});
     */
    async createImpAudienceGroup(impAudienceGroup) {
        const res = await this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/audienceGroup/imp`, Object.assign({}, impAudienceGroup));
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.updateAudienceGroupDescription} instead. Note: argument order changed and `audienceGroupId` is now a number.
     * @example
     * // Before:
     * client.setDescriptionAudienceGroup(description, audienceGroupId);
     * // After:
     * lineBotClient.updateAudienceGroupDescription(audienceGroupId, { description });
     */
    async setDescriptionAudienceGroup(description, audienceGroupId) {
        const res = await this.http.put(`${endpoints_js_1.MESSAGING_API_PREFIX}/audienceGroup/${audienceGroupId}/updateDescription`, {
            description,
        });
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.deleteAudienceGroup} instead. Note: `audienceGroupId` is now a number.
     * @example
     * // Before:
     * client.deleteAudienceGroup(audienceGroupId);
     * // After:
     * lineBotClient.deleteAudienceGroup(audienceGroupId);
     */
    async deleteAudienceGroup(audienceGroupId) {
        const res = await this.http.delete(`${endpoints_js_1.MESSAGING_API_PREFIX}/audienceGroup/${audienceGroupId}`);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.getAudienceData} instead. Note: `audienceGroupId` is now a number.
     * @example
     * // Before:
     * client.getAudienceGroup(audienceGroupId);
     * // After:
     * lineBotClient.getAudienceData(audienceGroupId);
     */
    async getAudienceGroup(audienceGroupId) {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/audienceGroup/${audienceGroupId}`);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.getAudienceGroups} instead. Note: the last two argument positions are swapped.
     * @example
     * // Before:
     * client.getAudienceGroups(page, description, status, size, createRoute, includesExternalPublicGroups);
     * // After:
     * lineBotClient.getAudienceGroups(page, description, status, size, includesExternalPublicGroups, createRoute);
     */
    async getAudienceGroups(page, description, status, size, createRoute, includesExternalPublicGroups) {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/audienceGroup/list`, {
            page,
            description,
            status,
            size,
            createRoute,
            includesExternalPublicGroups,
        });
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated No direct equivalent in LineBotClient. See docs/guide/migration.md for details.
     */
    async getAudienceGroupAuthorityLevel() {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/audienceGroup/authorityLevel`);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated No direct equivalent in LineBotClient. See docs/guide/migration.md for details.
     */
    async changeAudienceGroupAuthorityLevel(authorityLevel) {
        const res = await this.http.put(`${endpoints_js_1.MESSAGING_API_PREFIX}/audienceGroup/authorityLevel`, { authorityLevel });
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.getBotInfo} instead.
     * @example
     * // Before:
     * client.getBotInfo();
     * // After:
     * lineBotClient.getBotInfo();
     */
    async getBotInfo() {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/info`);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.setWebhookEndpoint} instead.
     * @example
     * // Before:
     * client.setWebhookEndpointUrl(endpoint);
     * // After:
     * lineBotClient.setWebhookEndpoint({ endpoint });
     */
    async setWebhookEndpointUrl(endpoint) {
        return this.http.put(`${endpoints_js_1.MESSAGING_API_PREFIX}/channel/webhook/endpoint`, { endpoint });
    }
    /**
     * @deprecated Use {@link LineBotClient.getWebhookEndpoint} instead.
     * @example
     * // Before:
     * client.getWebhookEndpointInfo();
     * // After:
     * lineBotClient.getWebhookEndpoint();
     */
    async getWebhookEndpointInfo() {
        const res = await this.http.get(`${endpoints_js_1.MESSAGING_API_PREFIX}/channel/webhook/endpoint`);
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.testWebhookEndpoint} instead.
     * @example
     * // Before:
     * client.testWebhookEndpoint(endpoint);
     * // After:
     * lineBotClient.testWebhookEndpoint({ endpoint });
     */
    async testWebhookEndpoint(endpoint) {
        const res = await this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/channel/webhook/test`, { endpoint });
        return (0, utils_js_1.ensureJSON)(res);
    }
    /**
     * @deprecated Use {@link LineBotClient.validateRichMenuObject} instead.
     * @example
     * // Before:
     * client.validateRichMenu(richMenu);
     * // After:
     * lineBotClient.validateRichMenuObject(richMenu);
     */
    async validateRichMenu(richMenu) {
        return await this.http.post(`${endpoints_js_1.MESSAGING_API_PREFIX}/richmenu/validate`, richMenu);
    }
}
exports.default = Client;
/**
 * @deprecated Use `channelAccessToken.ChannelAccessTokenClient` instead. See docs/guide/migration.md for the method mapping.
 */
class OAuth {
    constructor() {
        this.http = new http_axios_js_1.default();
    }
    /**
     * @deprecated Use {@link channelAccessToken.ChannelAccessTokenClient.issueChannelToken} instead.
     * @example
     * // Before:
     * oauth.issueAccessToken(client_id, client_secret);
     * // After:
     * const oauthClient = new channelAccessToken.ChannelAccessTokenClient({});
     * oauthClient.issueChannelToken('client_credentials', client_id, client_secret);
     */
    issueAccessToken(client_id, client_secret) {
        return this.http.postForm(`${endpoints_js_1.OAUTH_BASE_PREFIX}/accessToken`, {
            grant_type: "client_credentials",
            client_id,
            client_secret,
        });
    }
    /**
     * @deprecated Use {@link channelAccessToken.ChannelAccessTokenClient.revokeChannelToken} instead.
     * @example
     * // Before:
     * oauth.revokeAccessToken(access_token);
     * // After:
     * const oauthClient = new channelAccessToken.ChannelAccessTokenClient({});
     * oauthClient.revokeChannelToken(access_token);
     */
    revokeAccessToken(access_token) {
        return this.http.postForm(`${endpoints_js_1.OAUTH_BASE_PREFIX}/revoke`, { access_token });
    }
    /**
     * @deprecated Use {@link channelAccessToken.ChannelAccessTokenClient.verifyChannelTokenByJWT} instead.
     * @example
     * // Before:
     * oauth.verifyAccessToken(access_token);
     * // After:
     * const oauthClient = new channelAccessToken.ChannelAccessTokenClient({});
     * oauthClient.verifyChannelTokenByJWT(access_token);
     */
    verifyAccessToken(access_token) {
        return this.http.get(`${endpoints_js_1.OAUTH_BASE_PREFIX_V2_1}/verify`, { access_token });
    }
    /**
     * @deprecated No direct equivalent in LineBotClient. See docs/guide/migration.md for details.
     */
    verifyIdToken(id_token, client_id, nonce, user_id) {
        const body = {
            id_token,
            client_id,
        };
        if (nonce) {
            body.nonce = nonce;
        }
        if (user_id) {
            body.user_id = user_id;
        }
        return this.http.postForm(`${endpoints_js_1.OAUTH_BASE_PREFIX_V2_1}/verify`, body);
    }
    /**
     * @deprecated Use {@link channelAccessToken.ChannelAccessTokenClient.issueChannelTokenByJWT} instead.
     * @example
     * // Before:
     * oauth.issueChannelAccessTokenV2_1(client_assertion);
     * // After:
     * const oauthClient = new channelAccessToken.ChannelAccessTokenClient({});
     * oauthClient.issueChannelTokenByJWT('client_credentials', 'urn:ietf:params:oauth:client-assertion-type:jwt-bearer', client_assertion);
     */
    issueChannelAccessTokenV2_1(client_assertion) {
        return this.http.postForm(`${endpoints_js_1.OAUTH_BASE_PREFIX_V2_1}/token`, {
            grant_type: "client_credentials",
            client_assertion_type: "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
            client_assertion,
        });
    }
    /**
     * @deprecated Use {@link channelAccessToken.ChannelAccessTokenClient.getsAllValidChannelAccessTokenKeyIds} instead.
     * @example
     * // Before:
     * oauth.getChannelAccessTokenKeyIdsV2_1(client_assertion);
     * // After:
     * const oauthClient = new channelAccessToken.ChannelAccessTokenClient({});
     * oauthClient.getsAllValidChannelAccessTokenKeyIds('urn:ietf:params:oauth:client-assertion-type:jwt-bearer', client_assertion);
     */
    getChannelAccessTokenKeyIdsV2_1(client_assertion) {
        return this.http.get(`${endpoints_js_1.OAUTH_BASE_PREFIX_V2_1}/tokens/kid`, {
            client_assertion_type: "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
            client_assertion,
        });
    }
    /**
     * @deprecated Use {@link channelAccessToken.ChannelAccessTokenClient.revokeChannelTokenByJWT} instead.
     * @example
     * // Before:
     * oauth.revokeChannelAccessTokenV2_1(client_id, client_secret, access_token);
     * // After:
     * const oauthClient = new channelAccessToken.ChannelAccessTokenClient({});
     * oauthClient.revokeChannelTokenByJWT(client_id, client_secret, access_token);
     */
    revokeChannelAccessTokenV2_1(client_id, client_secret, access_token) {
        return this.http.postForm(`${endpoints_js_1.OAUTH_BASE_PREFIX_V2_1}/revoke`, {
            client_id,
            client_secret,
            access_token,
        });
    }
}
exports.OAuth = OAuth;
