// @line/bot-sdk/dist/cjs/exceptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HTTPFetchError = exports.HTTPError = exports.ReadError = exports.RequestError = exports.JSONParseError = exports.SignatureValidationFailed = void 0;
class SignatureValidationFailed extends Error {
    constructor(message, { signature } = {}) {
        super(message);
        this.name = this.constructor.name;
        Object.assign(this, { signature });
    }
}
exports.SignatureValidationFailed = SignatureValidationFailed;
class JSONParseError extends Error {
    constructor(message, { raw } = {}) {
        super(message);
        this.name = this.constructor.name;
        Object.assign(this, { raw });
    }
}
exports.JSONParseError = JSONParseError;
/**
 * @deprecated Legacy axios-based HTTP client only. `LineBotClient` (fetch-based) does not
 * throw this error. Network-level failures in `LineBotClient` surface as a native `TypeError`
 * from the Fetch API, not as `RequestError`.
 */
class RequestError extends Error {
    constructor(message, { code, originalError }) {
        super(message);
        this.name = this.constructor.name;
        Object.assign(this, { code, originalError });
    }
}
exports.RequestError = RequestError;
/**
 * @deprecated Legacy axios-based HTTP client only. `LineBotClient` (fetch-based) does not
 * throw this error. Read failures in `LineBotClient` propagate unwrapped from the Fetch API.
 */
class ReadError extends Error {
    constructor(message, { originalError }) {
        super(message);
        this.name = this.constructor.name;
        Object.assign(this, { originalError });
    }
}
exports.ReadError = ReadError;
/**
 * @deprecated Use `HTTPFetchError` instead. Thrown by the legacy axios-based HTTP client;
 * `LineBotClient` uses the Fetch API and throws `HTTPFetchError` on non-2xx responses.
 * Key differences: `HTTPFetchError` exposes `status`/`statusText`/`headers`/`body`
 * instead of `statusCode`/`statusMessage`/`originalError`.
 */
class HTTPError extends Error {
    constructor(message, { statusCode, statusMessage, originalError }) {
        super(message);
        this.name = this.constructor.name;
        Object.assign(this, { statusCode, statusMessage, originalError });
    }
}
exports.HTTPError = HTTPError;
class HTTPFetchError extends Error {
    constructor(message, { status, statusText, headers, body }) {
        super(message);
        this.name = this.constructor.name;
        Object.assign(this, { status, statusText, headers, body });
    }
}
exports.HTTPFetchError = HTTPFetchError;
