// openclaw/dist/extensions/mattermost/api.js
import { t as formatDocsLink } from "../../links-BFfjc3N-.js";
import { _ as normalizeAccountId, d as resolveThreadSessionKeys$1, g as DEFAULT_ACCOUNT_ID } from "../../session-key-BR3Z-ljs.js";
import { t as createDedupeCache } from "../../dedupe-CB5IJsQ1.js";
import { F as requireOpenAllowFrom, a as DmPolicySchema, c as GroupPolicySchema, m as MarkdownConfigSchema, n as BlockStreamingCoalesceSchema } from "../../zod-schema.core-BITC5-JP.js";
import { a as hasConfiguredSecretInput } from "../../types.secrets-BZdSA8i7.js";
import { r as buildChannelConfigSchema } from "../../config-schema-BEuKmAWr.js";
import { i as parseStrictPositiveInteger } from "../../parse-finite-number-HbZldjE_.js";
import { n as safeParseWithSchema, t as safeParseJsonWithSchema } from "../../zod-parse-SRMZ4WYD.js";
import { t as rawDataToString } from "../../ws-DLFbhvgw.js";
import { n as fetchWithSsrFGuard } from "../../fetch-guard-Bl48brXk.js";
import { c as createScopedChannelConfigAdapter, t as adaptScopedAccountAccessor } from "../../channel-config-helpers-CWYUF2eN.js";
import { n as isDangerousNameMatchingEnabled } from "../../dangerous-name-matching-CMg2IF_2.js";
import { n as resolveChannelGroupRequireMention } from "../../group-policy-D1X7pmp3.js";
import { n as describeAccountSnapshot } from "../../account-helpers-A6tF0W1f.js";
import { a as createSetupInputPresenceValidator, n as applySetupAccountConfigPatch, s as migrateBaseNameToDefaultAccount, t as applyAccountNameToChannelSection } from "../../setup-helpers-BiAtGxsL.js";
import { c as stripTargetKindPrefix, s as stripChannelTargetPrefix, t as buildChannelOutboundSessionRoute } from "../../core-D7mi2qmR.js";
import { t as normalizeOutboundThreadId } from "../../routing-DdBDhOmH.js";
import { r as buildSecretInputSchema } from "../../secret-input-D5U3kuko.js";
import { i as formatInboundFromLabel$1 } from "../../envelope-C2z9fFcf.js";
import { n as resolveControlCommandGate } from "../../command-gating-C6mMbL1P.js";
import { t as createAccountStatusSink } from "../../channel-lifecycle.core-CEzRKpfY.js";
import { n as formatNormalizedAllowFromEntries } from "../../allow-from-DjymPYUA.js";
import { a as warnMissingProviderGroupPolicyFallbackOnce, n as resolveAllowlistProviderRuntimeGroupPolicy, r as resolveDefaultGroupPolicy } from "../../runtime-group-policy-DxOE0SLn.js";
import { n as createRestrictSendersChannelSecurity, t as createDangerousNameMatchingMutableAllowlistWarningCollector } from "../../channel-policy-DIVRdPsQ.js";
import { n as readStoreAllowFromForDmPolicy, o as resolveDmGroupAccessWithLists, t as DM_GROUP_ACCESS_REASON } from "../../dm-policy-shared-CWGTUVOf.js";
import { a as buildPendingHistoryContextFromMap, s as clearHistoryEntriesIfEnabled, u as recordPendingHistoryEntryIfEnabled } from "../../history-ClGWPUk1.js";
import { t as createChannelDirectoryAdapter } from "../../directory-runtime-BrmKrim8.js";
import { n as logInboundDrop, r as logTypingFailure } from "../../logging-DomMbySE.js";
import { t as resolveChannelMediaMaxBytes } from "../../media-limits-bs8TnBXO.js";
import { f as createStandardChannelSetupStatus } from "../../setup-wizard-helpers-ecC16ic3.js";
import { t as createChannelReplyPipeline } from "../../channel-reply-pipeline-DkatqAK5.js";
import { n as createChannelPairingController, r as createLoggedPairingApprovalNotifier } from "../../channel-pairing-DrJTvhRN.js";
import { d as createDefaultChannelRuntimeState, u as createComputedAccountStatusAdapter } from "../../status-helpers-ChR3_7qO.js";
import { t as registerPluginHttpRoute } from "../../http-registry-UDbwC7Vv.js";
import "../../setup-Dp8bIdbL.js";
import "../../setup-runtime-QdMg-xhs.js";
import { n as resolveApprovalApprovers, t as createResolvedApproverActionAuthAdapter } from "../../approval-auth-helpers-DNJdxO4L.js";
import "../../config-runtime-OuR9WVXH.js";
import { t as buildAgentMediaPayload } from "../../agent-media-payload-t2IfQA3D.js";
import { a as isPrivateNetworkOptInEnabled, o as migrateLegacyFlatAllowPrivateNetworkAlias, r as hasLegacyFlatAllowPrivateNetworkAlias, u as ssrfPolicyFromPrivateNetworkOptIn } from "../../ssrf-policy-Cb9w9jMO.js";
import "../../ssrf-runtime-DGIvmaoK.js";
import { t as chunkTextForOutbound } from "../../text-chunking-BQ3u22Jv.js";
import { t as buildModelsProviderData } from "../../commands-models-BROMICu6.js";
import { t as listSkillCommandsForAgents } from "../../skill-commands-CwndRm6t.js";
import { n as buildPassiveProbedChannelStatusSummary } from "../../extension-shared-CKz43ndd.js";
import "../../channel-config-primitives-DiYud7LO.js";
import { t as createMessageToolButtonsSchema } from "../../channel-actions-DLDrCW4b.js";
import { n as createChatChannelPlugin } from "../../channel-core-BVR4R0_P.js";
import { t as zod_exports } from "../../zod-COH8D-AP.js";
import { B as fetchMattermostUserTeams, C as cleanupSlashCommands, D as resolveCallbackUrl, E as registerSlashCommands, F as MattermostPostSchema, H as readMattermostError, I as createMattermostClient, L as fetchMattermostChannel, M as resolveDefaultMattermostAccountId, N as resolveMattermostAccount, O as resolveSlashCommandConfig, P as resolveMattermostReplyToMode, R as fetchMattermostMe, S as DEFAULT_COMMAND_SPECS, T as isSlashCommandsEnabled, U as sendMattermostTyping, V as normalizeMattermostBaseUrl, W as updateMattermostPost, _ as buildMattermostAllowedModelRefs, a as setInteractionCallbackUrl, b as renderMattermostProviderPickerView, c as deactivateSlashCommands, d as sendMessageMattermost, f as resolveMattermostOpaqueTarget, g as normalizeMattermostAllowList, h as isMattermostSenderAllowed, i as resolveInteractionCallbackPath, j as listMattermostAccountIds, k as getMattermostRuntime, l as getSlashCommandState, m as authorizeMattermostCommandInvocation, n as computeInteractionCallbackUrl, o as setInteractionSecret, p as deliverMattermostReplyPayload, r as createMattermostInteractionHandler, s as activateSlashCommands, t as buildButtonProps, v as parseMattermostModelPickerContext, w as collectMattermostSlashCallbackPaths, x as resolveMattermostModelPickerCurrentModel, y as renderMattermostModelsPickerView, z as fetchMattermostUser } from "../../interactions-BrWHEMjm.js";
import { n as secretTargetRegistryEntries, t as collectRuntimeConfigAssignments } from "../../secret-contract-DzAqZdnf.js";
import WebSocket from "ws";
import { Type } from "@sinclair/typebox";
//#region extensions/mattermost/src/approval-auth.ts
const MATTERMOST_USER_ID_RE = /^[a-z0-9]{26}$/;
function normalizeMattermostApproverId(value) {
	const normalized = String(value).trim().replace(/^(mattermost|user):/i, "").replace(/^@/, "").trim().toLowerCase();
	return MATTERMOST_USER_ID_RE.test(normalized) ? normalized : void 0;
}
const mattermostApprovalAuth = createResolvedApproverActionAuthAdapter({
	channelLabel: "Mattermost",
	resolveApprovers: ({ cfg, accountId }) => {
		const account = resolveMattermostAccount({
			cfg,
			accountId
		}).config;
		return resolveApprovalApprovers({
			allowFrom: account.allowFrom,
			normalizeApprover: normalizeMattermostApproverId
		});
	},
	normalizeSenderId: (value) => normalizeMattermostApproverId(value)
});
//#endregion
//#region extensions/mattermost/src/config-schema-core.ts
const MattermostGroupSchema = zod_exports.z.object({ requireMention: zod_exports.z.boolean().optional() }).strict();
function requireMattermostOpenAllowFrom(params) {
	requireOpenAllowFrom({
		policy: params.policy,
		allowFrom: params.allowFrom,
		ctx: params.ctx,
		path: ["allowFrom"],
		message: "channels.mattermost.dmPolicy=\"open\" requires channels.mattermost.allowFrom to include \"*\""
	});
}
const DmChannelRetrySchema = zod_exports.z.object({
	maxRetries: zod_exports.z.number().int().min(0).max(10).optional(),
	initialDelayMs: zod_exports.z.number().int().min(100).max(6e4).optional(),
	maxDelayMs: zod_exports.z.number().int().min(1e3).max(6e4).optional(),
	timeoutMs: zod_exports.z.number().int().min(5e3).max(12e4).optional()
}).strict().refine((data) => {
	if (data.initialDelayMs !== void 0 && data.maxDelayMs !== void 0) return data.initialDelayMs <= data.maxDelayMs;
	return true;
}, {
	message: "initialDelayMs must be less than or equal to maxDelayMs",
	path: ["initialDelayMs"]
}).optional();
const MattermostSlashCommandsSchema = zod_exports.z.object({
	native: zod_exports.z.union([zod_exports.z.boolean(), zod_exports.z.literal("auto")]).optional(),
	nativeSkills: zod_exports.z.union([zod_exports.z.boolean(), zod_exports.z.literal("auto")]).optional(),
	callbackPath: zod_exports.z.string().optional(),
	callbackUrl: zod_exports.z.string().optional()
}).strict().optional();
const MattermostNetworkSchema = zod_exports.z.object({ dangerouslyAllowPrivateNetwork: zod_exports.z.boolean().optional() }).strict().optional();
const MattermostAccountSchemaBase = zod_exports.z.object({
	name: zod_exports.z.string().optional(),
	capabilities: zod_exports.z.array(zod_exports.z.string()).optional(),
	dangerouslyAllowNameMatching: zod_exports.z.boolean().optional(),
	markdown: MarkdownConfigSchema,
	enabled: zod_exports.z.boolean().optional(),
	configWrites: zod_exports.z.boolean().optional(),
	botToken: buildSecretInputSchema().optional(),
	baseUrl: zod_exports.z.string().optional(),
	chatmode: zod_exports.z.enum([
		"oncall",
		"onmessage",
		"onchar"
	]).optional(),
	oncharPrefixes: zod_exports.z.array(zod_exports.z.string()).optional(),
	requireMention: zod_exports.z.boolean().optional(),
	dmPolicy: DmPolicySchema.optional().default("pairing"),
	allowFrom: zod_exports.z.array(zod_exports.z.union([zod_exports.z.string(), zod_exports.z.number()])).optional(),
	groupAllowFrom: zod_exports.z.array(zod_exports.z.union([zod_exports.z.string(), zod_exports.z.number()])).optional(),
	groupPolicy: GroupPolicySchema.optional().default("allowlist"),
	textChunkLimit: zod_exports.z.number().int().positive().optional(),
	chunkMode: zod_exports.z.enum(["length", "newline"]).optional(),
	blockStreaming: zod_exports.z.boolean().optional(),
	blockStreamingCoalesce: BlockStreamingCoalesceSchema.optional(),
	replyToMode: zod_exports.z.enum([
		"off",
		"first",
		"all",
		"batched"
	]).optional(),
	responsePrefix: zod_exports.z.string().optional(),
	actions: zod_exports.z.object({ reactions: zod_exports.z.boolean().optional() }).optional(),
	commands: MattermostSlashCommandsSchema,
	interactions: zod_exports.z.object({
		callbackBaseUrl: zod_exports.z.string().optional(),
		allowedSourceIps: zod_exports.z.array(zod_exports.z.string()).optional()
	}).optional(),
	groups: zod_exports.z.record(zod_exports.z.string(), MattermostGroupSchema.optional()).optional(),
	network: MattermostNetworkSchema,
	dmChannelRetry: DmChannelRetrySchema
}).strict();
const MattermostAccountSchema = MattermostAccountSchemaBase.superRefine((value, ctx) => {
	requireMattermostOpenAllowFrom({
		policy: value.dmPolicy,
		allowFrom: value.allowFrom,
		ctx
	});
});
//#endregion
//#region extensions/mattermost/src/config-surface.ts
const MattermostChannelConfigSchema = buildChannelConfigSchema(MattermostAccountSchemaBase.extend({
	accounts: zod_exports.z.record(zod_exports.z.string(), MattermostAccountSchema.optional()).optional(),
	defaultAccount: zod_exports.z.string().optional()
}).superRefine((value, ctx) => {
	requireMattermostOpenAllowFrom({
		policy: value.dmPolicy,
		allowFrom: value.allowFrom,
		ctx
	});
}));
//#endregion
//#region extensions/mattermost/src/doctor.ts
function isRecord(value) {
	return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}
function isMattermostMutableAllowEntry(raw) {
	const text = raw.trim();
	if (!text || text === "*") return false;
	const normalized = text.replace(/^(mattermost|user):/i, "").replace(/^@/, "").trim().toLowerCase();
	if (/^[a-z0-9]{26}$/.test(normalized)) return false;
	return true;
}
const collectMattermostMutableAllowlistWarnings = createDangerousNameMatchingMutableAllowlistWarningCollector({
	channel: "mattermost",
	detector: isMattermostMutableAllowEntry,
	collectLists: (scope) => [{
		pathLabel: `${scope.prefix}.allowFrom`,
		list: scope.account.allowFrom
	}, {
		pathLabel: `${scope.prefix}.groupAllowFrom`,
		list: scope.account.groupAllowFrom
	}]
});
function hasLegacyMattermostAllowPrivateNetworkInAccounts(value) {
	const accounts = isRecord(value) ? value : null;
	return Boolean(accounts && Object.values(accounts).some((account) => hasLegacyFlatAllowPrivateNetworkAlias(isRecord(account) ? account : {})));
}
const MATTERMOST_LEGACY_CONFIG_RULES = [{
	path: ["channels", "mattermost"],
	message: "channels.mattermost.allowPrivateNetwork is legacy; use channels.mattermost.network.dangerouslyAllowPrivateNetwork instead. Run \"openclaw doctor --fix\".",
	match: (value) => hasLegacyFlatAllowPrivateNetworkAlias(isRecord(value) ? value : {})
}, {
	path: [
		"channels",
		"mattermost",
		"accounts"
	],
	message: "channels.mattermost.accounts.<id>.allowPrivateNetwork is legacy; use channels.mattermost.accounts.<id>.network.dangerouslyAllowPrivateNetwork instead. Run \"openclaw doctor --fix\".",
	match: hasLegacyMattermostAllowPrivateNetworkInAccounts
}];
function normalizeMattermostCompatibilityConfig(cfg) {
	const channels = isRecord(cfg.channels) ? cfg.channels : null;
	const mattermost = isRecord(channels?.mattermost) ? channels.mattermost : null;
	if (!mattermost) return {
		config: cfg,
		changes: []
	};
	const changes = [];
	let updatedMattermost = mattermost;
	let changed = false;
	const topLevel = migrateLegacyFlatAllowPrivateNetworkAlias({
		entry: updatedMattermost,
		pathPrefix: "channels.mattermost",
		changes
	});
	updatedMattermost = topLevel.entry;
	changed = changed || topLevel.changed;
	const accounts = isRecord(updatedMattermost.accounts) ? updatedMattermost.accounts : null;
	if (accounts) {
		let accountsChanged = false;
		const nextAccounts = { ...accounts };
		for (const [accountId, accountValue] of Object.entries(accounts)) {
			const account = isRecord(accountValue) ? accountValue : null;
			if (!account) continue;
			const migrated = migrateLegacyFlatAllowPrivateNetworkAlias({
				entry: account,
				pathPrefix: `channels.mattermost.accounts.${accountId}`,
				changes
			});
			if (!migrated.changed) continue;
			nextAccounts[accountId] = migrated.entry;
			accountsChanged = true;
		}
		if (accountsChanged) {
			updatedMattermost = {
				...updatedMattermost,
				accounts: nextAccounts
			};
			changed = true;
		}
	}
	if (!changed) return {
		config: cfg,
		changes: []
	};
	return {
		config: {
			...cfg,
			channels: {
				...cfg.channels,
				mattermost: updatedMattermost
			}
		},
		changes
	};
}
const mattermostDoctor = {
	legacyConfigRules: MATTERMOST_LEGACY_CONFIG_RULES,
	normalizeCompatibilityConfig: ({ cfg }) => normalizeMattermostCompatibilityConfig(cfg),
	collectMutableAllowlistWarnings: collectMattermostMutableAllowlistWarnings
};
//#endregion
//#region extensions/mattermost/src/group-mentions.ts
function resolveMattermostGroupRequireMention(params) {
	const account = resolveMattermostAccount({
		cfg: params.cfg,
		accountId: params.accountId
	});
	const requireMentionOverride = typeof params.requireMentionOverride === "boolean" ? params.requireMentionOverride : account.requireMention;
	return resolveChannelGroupRequireMention({
		cfg: params.cfg,
		channel: "mattermost",
		groupId: params.groupId,
		accountId: params.accountId,
		requireMentionOverride
	});
}
//#endregion
//#region extensions/mattermost/src/mattermost/directory.ts
function buildClient(params) {
	const account = resolveMattermostAccount({
		cfg: params.cfg,
		accountId: params.accountId
	});
	if (!account.enabled || !account.botToken || !account.baseUrl) return null;
	return createMattermostClient({
		baseUrl: account.baseUrl,
		botToken: account.botToken,
		allowPrivateNetwork: isPrivateNetworkOptInEnabled(account.config)
	});
}
/**
* Build clients from ALL enabled accounts (deduplicated by token).
*
* We always scan every account because:
* - Private channels are only visible to bots that are members
* - The requesting agent's account may have an expired/invalid token
*
* This means a single healthy bot token is enough for directory discovery.
*/
function buildClients(params) {
	const accountIds = listMattermostAccountIds(params.cfg);
	const seen = /* @__PURE__ */ new Set();
	const clients = [];
	for (const id of accountIds) {
		const client = buildClient({
			cfg: params.cfg,
			accountId: id
		});
		if (client && !seen.has(client.token)) {
			seen.add(client.token);
			clients.push(client);
		}
	}
	return clients;
}
/**
* List channels (public + private) visible to any configured bot account.
*
* NOTE: Uses per_page=200 which covers most instances. Mattermost does not
* return a "has more" indicator, so very large instances (200+ channels per bot)
* may see incomplete results. Pagination can be added if needed.
*/
async function listMattermostDirectoryGroups(params) {
	const clients = buildClients(params);
	if (!clients.length) return [];
	const q = params.query?.trim().toLowerCase() || "";
	const seenIds = /* @__PURE__ */ new Set();
	const entries = [];
	for (const client of clients) try {
		const me = await fetchMattermostMe(client);
		const channels = await client.request(`/users/${me.id}/channels?per_page=200`);
		for (const ch of channels) {
			if (ch.type !== "O" && ch.type !== "P") continue;
			if (seenIds.has(ch.id)) continue;
			if (q) {
				const name = (ch.name ?? "").toLowerCase();
				const display = (ch.display_name ?? "").toLowerCase();
				if (!name.includes(q) && !display.includes(q)) continue;
			}
			seenIds.add(ch.id);
			entries.push({
				kind: "group",
				id: `channel:${ch.id}`,
				name: ch.name ?? void 0,
				handle: ch.display_name ?? void 0
			});
		}
	} catch (err) {
		console.debug?.("[mattermost-directory] listGroups: skipping account:", err?.message);
		continue;
	}
	return params.limit && params.limit > 0 ? entries.slice(0, params.limit) : entries;
}
/**
* List team members as peer directory entries.
*
* Uses only the first available client since all bots in a team see the same
* user list (unlike channels where membership varies). Uses the first team
* returned — multi-team setups will only see members from that team.
*
* NOTE: per_page=200 for member listing; same pagination caveat as groups.
*/
async function listMattermostDirectoryPeers(params) {
	const clients = buildClients(params);
	if (!clients.length) return [];
	const client = clients[0];
	try {
		const me = await fetchMattermostMe(client);
		const teams = await client.request("/users/me/teams");
		if (!teams.length) return [];
		const teamId = teams[0].id;
		const q = params.query?.trim().toLowerCase() || "";
		let users;
		if (q) users = await client.request("/users/search", {
			method: "POST",
			body: JSON.stringify({
				term: q,
				team_id: teamId
			})
		});
		else {
			const userIds = (await client.request(`/teams/${teamId}/members?per_page=200`)).map((m) => m.user_id).filter((id) => id !== me.id);
			if (!userIds.length) return [];
			users = await client.request("/users/ids", {
				method: "POST",
				body: JSON.stringify(userIds)
			});
		}
		const entries = users.filter((u) => u.id !== me.id).map((u) => ({
			kind: "user",
			id: `user:${u.id}`,
			name: u.username ?? void 0,
			handle: [u.first_name, u.last_name].filter(Boolean).join(" ").trim() || u.nickname || void 0
		}));
		return params.limit && params.limit > 0 ? entries.slice(0, params.limit) : entries;
	} catch (err) {
		console.debug?.("[mattermost-directory] listPeers failed:", err?.message);
		return [];
	}
}
//#endregion
//#region extensions/mattermost/src/mattermost/monitor-gating.ts
function mapMattermostChannelTypeToChatType(channelType) {
	if (!channelType) return "channel";
	const normalized = channelType.trim().toUpperCase();
	if (normalized === "D") return "direct";
	if (normalized === "G" || normalized === "P") return "group";
	return "channel";
}
function evaluateMattermostMentionGate(params) {
	const shouldRequireMention = params.kind !== "direct" && params.resolveRequireMention({
		cfg: params.cfg,
		channel: "mattermost",
		accountId: params.accountId,
		groupId: params.channelId,
		requireMentionOverride: params.requireMentionOverride
	});
	const shouldBypassMention = params.isControlCommand && shouldRequireMention && !params.wasMentioned && params.commandAuthorized;
	const effectiveWasMentioned = params.wasMentioned || shouldBypassMention || params.oncharTriggered;
	if (params.oncharEnabled && !params.oncharTriggered && !params.wasMentioned && !params.isControlCommand) return {
		shouldRequireMention,
		shouldBypassMention,
		effectiveWasMentioned,
		dropReason: "onchar-not-triggered"
	};
	if (params.kind !== "direct" && shouldRequireMention && params.canDetectMention && !effectiveWasMentioned) return {
		shouldRequireMention,
		shouldBypassMention,
		effectiveWasMentioned,
		dropReason: "missing-mention"
	};
	return {
		shouldRequireMention,
		shouldBypassMention,
		effectiveWasMentioned,
		dropReason: null
	};
}
//#endregion
//#region extensions/mattermost/src/mattermost/monitor-helpers.ts
const formatInboundFromLabel = formatInboundFromLabel$1;
function resolveThreadSessionKeys(params) {
	return resolveThreadSessionKeys$1({
		...params,
		normalizeThreadId: (threadId) => threadId
	});
}
/**
* Strip bot mention from message text while preserving newlines and
* block-level Markdown formatting (headings, lists, blockquotes).
*/
function normalizeMention(text, mention) {
	if (!mention) return text.trim();
	const escaped = mention.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
	const hasMentionRe = new RegExp(`@${escaped}\\b`, "i");
	const leadingMentionRe = new RegExp(`^([\\t ]*)@${escaped}\\b[\\t ]*`, "i");
	const trailingMentionRe = new RegExp(`[\\t ]*@${escaped}\\b[\\t ]*$`, "i");
	const normalizedLines = text.split("\n").map((line) => {
		const hadMention = hasMentionRe.test(line);
		const normalizedLine = line.replace(leadingMentionRe, "$1").replace(trailingMentionRe, "").replace(new RegExp(`@${escaped}\\b`, "gi"), "").replace(/(\S)[ \t]{2,}/g, "$1 ");
		return {
			text: normalizedLine,
			mentionOnlyBlank: hadMention && normalizedLine.trim() === ""
		};
	});
	while (normalizedLines[0]?.mentionOnlyBlank) normalizedLines.shift();
	while (normalizedLines.at(-1)?.text.trim() === "") normalizedLines.pop();
	return normalizedLines.map((line) => line.text).join("\n");
}
//#endregion
//#region extensions/mattermost/src/mattermost/monitor-onchar.ts
const DEFAULT_ONCHAR_PREFIXES = [">", "!"];
function resolveOncharPrefixes(prefixes) {
	const cleaned = prefixes?.map((entry) => entry.trim()).filter(Boolean) ?? DEFAULT_ONCHAR_PREFIXES;
	return cleaned.length > 0 ? cleaned : DEFAULT_ONCHAR_PREFIXES;
}
function stripOncharPrefix(text, prefixes) {
	const trimmed = text.trimStart();
	for (const prefix of prefixes) {
		if (!prefix) continue;
		if (trimmed.startsWith(prefix)) return {
			triggered: true,
			stripped: trimmed.slice(prefix.length).trimStart()
		};
	}
	return {
		triggered: false,
		stripped: text
	};
}
//#endregion
//#region extensions/mattermost/src/mattermost/monitor-resources.ts
const CHANNEL_CACHE_TTL_MS = 5 * 6e4;
const USER_CACHE_TTL_MS = 10 * 6e4;
function createMattermostMonitorResources(params) {
	const { accountId, callbackUrl, client, logger, mediaMaxBytes, fetchRemoteMedia, saveMediaBuffer, mediaKindFromMime } = params;
	const channelCache = /* @__PURE__ */ new Map();
	const userCache = /* @__PURE__ */ new Map();
	const resolveMattermostMedia = async (fileIds) => {
		const ids = (fileIds ?? []).map((id) => id?.trim()).filter(Boolean);
		if (ids.length === 0) return [];
		const out = [];
		for (const fileId of ids) try {
			const fetched = await fetchRemoteMedia({
				url: `${client.apiBaseUrl}/files/${fileId}`,
				requestInit: { headers: { Authorization: `Bearer ${client.token}` } },
				filePathHint: fileId,
				maxBytes: mediaMaxBytes,
				ssrfPolicy: { allowedHostnames: [new URL(client.baseUrl).hostname] }
			});
			const saved = await saveMediaBuffer(Buffer.from(fetched.buffer), fetched.contentType ?? void 0, "inbound", mediaMaxBytes);
			const contentType = saved.contentType ?? fetched.contentType ?? void 0;
			out.push({
				path: saved.path,
				contentType,
				kind: mediaKindFromMime(contentType) ?? "unknown"
			});
		} catch (err) {
			logger.debug?.(`mattermost: failed to download file ${fileId}: ${String(err)}`);
		}
		return out;
	};
	const sendTypingIndicator = async (channelId, parentId) => {
		await sendMattermostTyping(client, {
			channelId,
			parentId
		});
	};
	const resolveChannelInfo = async (channelId) => {
		const cached = channelCache.get(channelId);
		if (cached && cached.expiresAt > Date.now()) return cached.value;
		try {
			const info = await fetchMattermostChannel(client, channelId);
			channelCache.set(channelId, {
				value: info,
				expiresAt: Date.now() + CHANNEL_CACHE_TTL_MS
			});
			return info;
		} catch (err) {
			logger.debug?.(`mattermost: channel lookup failed: ${String(err)}`);
			channelCache.set(channelId, {
				value: null,
				expiresAt: Date.now() + CHANNEL_CACHE_TTL_MS
			});
			return null;
		}
	};
	const resolveUserInfo = async (userId) => {
		const cached = userCache.get(userId);
		if (cached && cached.expiresAt > Date.now()) return cached.value;
		try {
			const info = await fetchMattermostUser(client, userId);
			userCache.set(userId, {
				value: info,
				expiresAt: Date.now() + USER_CACHE_TTL_MS
			});
			return info;
		} catch (err) {
			logger.debug?.(`mattermost: user lookup failed: ${String(err)}`);
			userCache.set(userId, {
				value: null,
				expiresAt: Date.now() + USER_CACHE_TTL_MS
			});
			return null;
		}
	};
	const buildModelPickerProps = (channelId, buttons) => buildButtonProps({
		callbackUrl,
		accountId,
		channelId,
		buttons
	});
	const updateModelPickerPost = async (params) => {
		const props = buildModelPickerProps(params.channelId, params.buttons ?? []) ?? { attachments: [] };
		await updateMattermostPost(client, params.postId, {
			message: params.message,
			props
		});
		return {};
	};
	return {
		resolveMattermostMedia,
		sendTypingIndicator,
		resolveChannelInfo,
		resolveUserInfo,
		updateModelPickerPost
	};
}
//#endregion
//#region extensions/mattermost/src/mattermost/monitor-slash.ts
function isLoopbackHost$1(hostname) {
	return hostname === "localhost" || hostname === "127.0.0.1" || hostname === "::1";
}
function buildSlashCommands(params) {
	const commandsToRegister = [...DEFAULT_COMMAND_SPECS];
	if (!params.nativeSkills) return commandsToRegister;
	try {
		const skillCommands = listSkillCommandsForAgents({ cfg: params.cfg });
		for (const spec of skillCommands) {
			const name = typeof spec.name === "string" ? spec.name.trim() : "";
			if (!name) continue;
			const trigger = name.startsWith("oc_") ? name : `oc_${name}`;
			commandsToRegister.push({
				trigger,
				description: spec.description || `Run skill ${name}`,
				autoComplete: true,
				autoCompleteHint: "[args]",
				originalName: name
			});
		}
	} catch (err) {
		params.runtime.error?.(`mattermost: failed to list skill commands: ${String(err)}`);
	}
	return commandsToRegister;
}
function dedupeSlashCommands(commands) {
	const seen = /* @__PURE__ */ new Set();
	return commands.filter((cmd) => {
		const key = cmd.trigger.trim();
		if (!key || seen.has(key)) return false;
		seen.add(key);
		return true;
	});
}
function buildTriggerMap(commands) {
	const triggerMap = /* @__PURE__ */ new Map();
	for (const cmd of commands) if (cmd.originalName) triggerMap.set(cmd.trigger, cmd.originalName);
	return triggerMap;
}
function warnOnSuspiciousCallbackUrl(params) {
	try {
		const mmHost = new URL(normalizeMattermostBaseUrl(params.baseUrl) ?? params.baseUrl).hostname;
		const callbackHost = new URL(params.callbackUrl).hostname;
		if (isLoopbackHost$1(callbackHost) && !isLoopbackHost$1(mmHost)) params.runtime.error?.(`mattermost: slash commands callbackUrl resolved to ${params.callbackUrl} (loopback) while baseUrl is ${params.baseUrl}. This MAY be unreachable depending on your deployment. If native slash commands don't work, set channels.mattermost.commands.callbackUrl to a URL reachable from the Mattermost server (e.g. your public reverse proxy URL).`);
	} catch {}
}
async function registerSlashCommandsAcrossTeams(params) {
	const registered = [];
	let teamRegistrationFailures = 0;
	for (const team of params.teams) try {
		const created = await registerSlashCommands({
			client: params.client,
			teamId: team.id,
			creatorUserId: params.botUserId,
			callbackUrl: params.callbackUrl,
			commands: params.commands,
			log: (msg) => params.runtime.log?.(msg)
		});
		registered.push(...created);
	} catch (err) {
		teamRegistrationFailures += 1;
		params.runtime.error?.(`mattermost: failed to register slash commands for team ${team.id}: ${String(err)}`);
	}
	return {
		registered,
		teamRegistrationFailures
	};
}
async function registerMattermostMonitorSlashCommands(params) {
	const commandsRaw = params.account.config.commands;
	const slashConfig = resolveSlashCommandConfig(commandsRaw);
	if (!isSlashCommandsEnabled(slashConfig)) return;
	try {
		const teams = await fetchMattermostUserTeams(params.client, params.botUserId);
		const slashCallbackUrl = resolveCallbackUrl({
			config: slashConfig,
			gatewayPort: parseStrictPositiveInteger(process.env.OPENCLAW_GATEWAY_PORT?.trim()) ?? params.cfg.gateway?.port ?? 18789,
			gatewayHost: params.cfg.gateway?.customBindHost ?? void 0
		});
		warnOnSuspiciousCallbackUrl({
			runtime: params.runtime,
			baseUrl: params.baseUrl,
			callbackUrl: slashCallbackUrl
		});
		const dedupedCommands = dedupeSlashCommands(buildSlashCommands({
			cfg: params.cfg,
			runtime: params.runtime,
			nativeSkills: slashConfig.nativeSkills === true
		}));
		const { registered, teamRegistrationFailures } = await registerSlashCommandsAcrossTeams({
			client: params.client,
			teams,
			botUserId: params.botUserId,
			callbackUrl: slashCallbackUrl,
			commands: dedupedCommands,
			runtime: params.runtime
		});
		if (registered.length === 0) {
			params.runtime.error?.("mattermost: native slash commands enabled but no commands could be registered; keeping slash callbacks inactive");
			return;
		}
		if (teamRegistrationFailures > 0) params.runtime.error?.(`mattermost: slash command registration completed with ${teamRegistrationFailures} team error(s)`);
		activateSlashCommands({
			account: params.account,
			commandTokens: registered.map((cmd) => cmd.token).filter(Boolean),
			registeredCommands: registered,
			triggerMap: buildTriggerMap(dedupedCommands),
			api: {
				cfg: params.cfg,
				runtime: params.runtime
			},
			log: (msg) => params.runtime.log?.(msg)
		});
		params.runtime.log?.(`mattermost: slash commands registered (${registered.length} commands across ${teams.length} teams, callback=${slashCallbackUrl})`);
	} catch (err) {
		params.runtime.error?.(`mattermost: failed to register slash commands: ${String(err)}`);
	}
}
//#endregion
//#region extensions/mattermost/src/mattermost/monitor-websocket.ts
const MattermostEventPayloadSchema = zod_exports.z.object({
	event: zod_exports.z.string().optional(),
	data: zod_exports.z.object({
		post: zod_exports.z.union([zod_exports.z.string(), MattermostPostSchema]).optional(),
		reaction: zod_exports.z.union([zod_exports.z.string(), zod_exports.z.record(zod_exports.z.string(), zod_exports.z.unknown())]).optional(),
		channel_id: zod_exports.z.string().optional(),
		channel_name: zod_exports.z.string().optional(),
		channel_display_name: zod_exports.z.string().optional(),
		channel_type: zod_exports.z.string().optional(),
		sender_name: zod_exports.z.string().optional(),
		team_id: zod_exports.z.string().optional()
	}).optional(),
	broadcast: zod_exports.z.object({
		channel_id: zod_exports.z.string().optional(),
		team_id: zod_exports.z.string().optional(),
		user_id: zod_exports.z.string().optional()
	}).optional()
});
function parseMattermostEventPayload(raw) {
	return safeParseJsonWithSchema(MattermostEventPayloadSchema, raw);
}
function parseMattermostPost(value) {
	if (typeof value === "string") return safeParseJsonWithSchema(MattermostPostSchema, value);
	return safeParseWithSchema(MattermostPostSchema, value);
}
var WebSocketClosedBeforeOpenError = class extends Error {
	constructor(code, reason) {
		super(`websocket closed before open (code ${code})`);
		this.code = code;
		this.reason = reason;
		this.name = "WebSocketClosedBeforeOpenError";
	}
};
const defaultMattermostWebSocketFactory = (url) => new WebSocket(url);
function parsePostedPayload(payload) {
	if (payload.event !== "posted") return null;
	const postData = payload.data?.post;
	if (!postData) return null;
	const post = parseMattermostPost(postData);
	if (!post) return null;
	return {
		payload,
		post
	};
}
function createMattermostConnectOnce(opts) {
	const webSocketFactory = opts.webSocketFactory ?? defaultMattermostWebSocketFactory;
	const healthCheckIntervalMs = opts.healthCheckIntervalMs ?? 3e4;
	return async () => {
		const ws = webSocketFactory(opts.wsUrl);
		const onAbort = () => ws.terminate();
		opts.abortSignal?.addEventListener("abort", onAbort, { once: true });
		const getBotUpdateAt = opts.getBotUpdateAt;
		try {
			return await new Promise((resolve, reject) => {
				let opened = false;
				let settled = false;
				let healthCheckEnabled = getBotUpdateAt != null;
				let healthCheckInFlight = false;
				let healthCheckTimer;
				let initialUpdateAt;
				const clearTimers = () => {
					if (healthCheckTimer !== void 0) {
						clearTimeout(healthCheckTimer);
						healthCheckTimer = void 0;
					}
				};
				const stopHealthChecks = () => {
					healthCheckEnabled = false;
					clearTimers();
				};
				const scheduleHealthCheck = () => {
					if (!getBotUpdateAt || !healthCheckEnabled || settled || healthCheckInFlight) return;
					healthCheckTimer = setTimeout(() => {
						healthCheckTimer = void 0;
						runHealthCheck();
					}, healthCheckIntervalMs);
				};
				const runHealthCheck = async () => {
					if (!getBotUpdateAt || !healthCheckEnabled || settled || healthCheckInFlight) return;
					healthCheckInFlight = true;
					try {
						const current = await getBotUpdateAt();
						if (!healthCheckEnabled || settled) return;
						if (initialUpdateAt === void 0) {
							initialUpdateAt = current;
							return;
						}
						if (current !== initialUpdateAt) {
							opts.runtime.log?.(`mattermost: bot account updated (update_at changed: ${initialUpdateAt} → ${current}) — reconnecting`);
							stopHealthChecks();
							ws.terminate();
						}
					} catch (err) {
						if (!healthCheckEnabled || settled) return;
						const label = initialUpdateAt === void 0 ? "mattermost: failed to get initial update_at" : "mattermost: health check error";
						opts.runtime.error?.(`${label}: ${String(err)}`);
					} finally {
						healthCheckInFlight = false;
						scheduleHealthCheck();
					}
				};
				const resolveOnce = () => {
					if (settled) return;
					settled = true;
					stopHealthChecks();
					resolve();
				};
				const rejectOnce = (error) => {
					if (settled) return;
					settled = true;
					stopHealthChecks();
					reject(error);
				};
				ws.on("open", () => {
					opened = true;
					opts.statusSink?.({
						connected: true,
						lastConnectedAt: Date.now(),
						lastError: null
					});
					ws.send(JSON.stringify({
						seq: opts.nextSeq(),
						action: "authentication_challenge",
						data: { token: opts.botToken }
					}));
					if (getBotUpdateAt) runHealthCheck();
				});
				ws.on("message", async (data) => {
					const payload = parseMattermostEventPayload(rawDataToString(data));
					if (!payload) return;
					if (payload.event === "reaction_added" || payload.event === "reaction_removed") {
						if (!opts.onReaction) return;
						try {
							await opts.onReaction(payload);
						} catch (err) {
							opts.runtime.error?.(`mattermost reaction handler failed: ${String(err)}`);
						}
						return;
					}
					if (payload.event !== "posted") return;
					const parsed = parsePostedPayload(payload);
					if (!parsed) return;
					try {
						await opts.onPosted(parsed.post, parsed.payload);
					} catch (err) {
						opts.runtime.error?.(`mattermost handler failed: ${String(err)}`);
					}
				});
				ws.on("close", (code, reason) => {
					stopHealthChecks();
					const message = reasonToString(reason);
					opts.statusSink?.({
						connected: false,
						lastDisconnect: {
							at: Date.now(),
							status: code,
							error: message || void 0
						}
					});
					if (opened) {
						resolveOnce();
						return;
					}
					rejectOnce(new WebSocketClosedBeforeOpenError(code, message || void 0));
				});
				ws.on("error", (err) => {
					opts.runtime.error?.(`mattermost websocket error: ${String(err)}`);
					opts.statusSink?.({ lastError: String(err) });
					try {
						ws.close();
					} catch {}
				});
			});
		} finally {
			opts.abortSignal?.removeEventListener("abort", onAbort);
		}
	};
}
function reasonToString(reason) {
	if (!reason) return "";
	if (typeof reason === "string") return reason;
	return reason.length > 0 ? reason.toString("utf8") : "";
}
//#endregion
//#region extensions/mattermost/src/mattermost/reconnect.ts
/**
* Reconnection loop with exponential backoff.
*
* Calls `connectFn` in a while loop. On normal resolve (connection closed),
* the backoff resets. On thrown error (connection failed), the current delay is
* used, then doubled for the next retry.
* The loop exits when `abortSignal` fires.
*/
async function runWithReconnect(connectFn, opts = {}) {
	const { initialDelayMs = 2e3, maxDelayMs = 6e4 } = opts;
	const jitterRatio = Math.max(0, opts.jitterRatio ?? 0);
	const random = opts.random ?? Math.random;
	let retryDelay = initialDelayMs;
	let attempt = 0;
	while (!opts.abortSignal?.aborted) {
		let shouldIncreaseDelay = false;
		let outcome = "resolved";
		let error;
		try {
			await connectFn();
			retryDelay = initialDelayMs;
		} catch (err) {
			if (opts.abortSignal?.aborted) return;
			outcome = "rejected";
			error = err;
			opts.onError?.(err);
			shouldIncreaseDelay = true;
		}
		if (opts.abortSignal?.aborted) return;
		const delayMs = withJitter(retryDelay, jitterRatio, random);
		if (!(opts.shouldReconnect?.({
			attempt,
			delayMs,
			outcome,
			error
		}) ?? true)) return;
		opts.onReconnect?.(delayMs);
		await sleepAbortable(delayMs, opts.abortSignal);
		if (shouldIncreaseDelay) retryDelay = Math.min(retryDelay * 2, maxDelayMs);
		attempt++;
	}
}
function withJitter(baseMs, jitterRatio, random) {
	if (jitterRatio <= 0) return baseMs;
	const normalized = Math.max(0, Math.min(1, random()));
	const spread = baseMs * jitterRatio;
	return Math.max(1, Math.round(baseMs - spread + normalized * spread * 2));
}
function sleepAbortable(ms, signal) {
	return new Promise((resolve) => {
		if (signal?.aborted) {
			resolve();
			return;
		}
		const onAbort = () => {
			clearTimeout(timer);
			resolve();
		};
		const timer = setTimeout(() => {
			signal?.removeEventListener("abort", onAbort);
			resolve();
		}, ms);
		signal?.addEventListener("abort", onAbort, { once: true });
	});
}
//#endregion
//#region extensions/mattermost/src/mattermost/monitor.ts
const RECENT_MATTERMOST_MESSAGE_TTL_MS = 5 * 6e4;
const RECENT_MATTERMOST_MESSAGE_MAX = 2e3;
function isLoopbackHost(hostname) {
	return hostname === "localhost" || hostname === "127.0.0.1" || hostname === "::1";
}
function normalizeInteractionSourceIps(values) {
	return (values ?? []).map((value) => value.trim()).filter(Boolean);
}
const recentInboundMessages = createDedupeCache({
	ttlMs: RECENT_MATTERMOST_MESSAGE_TTL_MS,
	maxSize: RECENT_MATTERMOST_MESSAGE_MAX
});
function resolveRuntime(opts) {
	return opts.runtime ?? {
		log: console.log,
		error: console.error,
		exit: (code) => {
			throw new Error(`exit ${code}`);
		}
	};
}
function isSystemPost(post) {
	const type = post.type?.trim();
	return Boolean(type);
}
function channelChatType(kind) {
	if (kind === "direct") return "direct";
	if (kind === "group") return "group";
	return "channel";
}
function resolveMattermostReplyRootId(params) {
	const threadRootId = params.threadRootId?.trim();
	if (threadRootId) return threadRootId;
	return params.replyToId?.trim() || void 0;
}
function resolveMattermostEffectiveReplyToId(params) {
	const threadRootId = params.threadRootId?.trim();
	if (threadRootId && params.replyToMode !== "off") return threadRootId;
	if (params.kind === "direct") return;
	const postId = params.postId?.trim();
	if (!postId) return;
	return params.replyToMode === "all" || params.replyToMode === "first" || params.replyToMode === "batched" ? postId : void 0;
}
function resolveMattermostThreadSessionContext(params) {
	const effectiveReplyToId = resolveMattermostEffectiveReplyToId({
		kind: params.kind,
		postId: params.postId,
		replyToMode: params.replyToMode,
		threadRootId: params.threadRootId
	});
	const threadKeys = resolveThreadSessionKeys({
		baseSessionKey: params.baseSessionKey,
		threadId: effectiveReplyToId,
		parentSessionKey: effectiveReplyToId ? params.baseSessionKey : void 0
	});
	return {
		effectiveReplyToId,
		sessionKey: threadKeys.sessionKey,
		parentSessionKey: threadKeys.parentSessionKey
	};
}
function resolveMattermostReactionChannelId(payload) {
	return payload.broadcast?.channel_id?.trim() || payload.data?.channel_id?.trim() || void 0;
}
function buildMattermostAttachmentPlaceholder(mediaList) {
	if (mediaList.length === 0) return "";
	if (mediaList.length === 1) return `<media:${mediaList[0].kind === "unknown" ? "document" : mediaList[0].kind}>`;
	const allImages = mediaList.every((media) => media.kind === "image");
	const label = allImages ? "image" : "file";
	const suffix = mediaList.length === 1 ? label : `${label}s`;
	return `${allImages ? "<media:image>" : "<media:document>"} (${mediaList.length} ${suffix})`;
}
function buildMattermostWsUrl(baseUrl) {
	const normalized = normalizeMattermostBaseUrl(baseUrl);
	if (!normalized) throw new Error("Mattermost baseUrl is required");
	return `${normalized.replace(/^http/i, "ws")}/api/v4/websocket`;
}
async function monitorMattermostProvider(opts = {}) {
	const core = getMattermostRuntime();
	const runtime = resolveRuntime(opts);
	const cfg = opts.config ?? core.config.loadConfig();
	const account = resolveMattermostAccount({
		cfg,
		accountId: opts.accountId
	});
	const pairing = createChannelPairingController({
		core,
		channel: "mattermost",
		accountId: account.accountId
	});
	const allowNameMatching = isDangerousNameMatchingEnabled(account.config);
	const botToken = opts.botToken?.trim() || account.botToken?.trim();
	if (!botToken) throw new Error(`Mattermost bot token missing for account "${account.accountId}" (set channels.mattermost.accounts.${account.accountId}.botToken or MATTERMOST_BOT_TOKEN for default).`);
	const baseUrl = normalizeMattermostBaseUrl(opts.baseUrl ?? account.baseUrl);
	if (!baseUrl) throw new Error(`Mattermost baseUrl missing for account "${account.accountId}" (set channels.mattermost.accounts.${account.accountId}.baseUrl or MATTERMOST_URL for default).`);
	const client = createMattermostClient({
		baseUrl,
		botToken,
		allowPrivateNetwork: isPrivateNetworkOptInEnabled(account.config)
	});
	let botUser;
	await runWithReconnect(async () => {
		botUser = await fetchMattermostMe(client);
	}, {
		abortSignal: opts.abortSignal,
		jitterRatio: .2,
		shouldReconnect: ({ outcome }) => outcome === "rejected",
		onError: (err) => {
			runtime.error?.(`mattermost: API auth failed: ${String(err)}`);
			opts.statusSink?.({
				lastError: String(err),
				connected: false
			});
		},
		onReconnect: (delayMs) => {
			runtime.log?.(`mattermost: API not accessible, retrying in ${Math.round(delayMs / 1e3)}s`);
		}
	});
	if (opts.abortSignal?.aborted) return;
	const botUserId = botUser.id;
	const botUsername = botUser.username?.trim() || void 0;
	runtime.log?.(`mattermost connected as ${botUsername ? `@${botUsername}` : botUserId}`);
	await registerMattermostMonitorSlashCommands({
		client,
		cfg,
		runtime,
		account,
		baseUrl,
		botUserId
	});
	const slashEnabled = getSlashCommandState(account.accountId) != null;
	setInteractionSecret(account.accountId, botToken);
	const interactionPath = resolveInteractionCallbackPath(account.accountId);
	const callbackUrl = computeInteractionCallbackUrl(account.accountId, {
		gateway: cfg.gateway,
		interactions: account.config.interactions
	});
	setInteractionCallbackUrl(account.accountId, callbackUrl);
	const allowedInteractionSourceIps = normalizeInteractionSourceIps(account.config.interactions?.allowedSourceIps);
	try {
		const mmHost = new URL(baseUrl).hostname;
		const callbackHost = new URL(callbackUrl).hostname;
		if (isLoopbackHost(callbackHost) && !isLoopbackHost(mmHost)) runtime.error?.(`mattermost: interactions callbackUrl resolved to ${callbackUrl} (loopback) while baseUrl is ${baseUrl}. This MAY be unreachable depending on your deployment. If button clicks don't work, set channels.mattermost.interactions.callbackBaseUrl to a URL reachable from the Mattermost server (e.g. your public reverse proxy URL).`);
		if (!isLoopbackHost(callbackHost) && allowedInteractionSourceIps.length === 0) runtime.error?.(`mattermost: interactions callbackUrl resolved to ${callbackUrl} without channels.mattermost.interactions.allowedSourceIps. For safety, non-loopback callback sources will be rejected until you allowlist the Mattermost server or trusted ingress IPs.`);
	} catch {}
	const effectiveInteractionSourceIps = allowedInteractionSourceIps.length > 0 ? allowedInteractionSourceIps : ["127.0.0.1", "::1"];
	const unregisterInteractions = registerPluginHttpRoute({
		path: interactionPath,
		fallbackPath: "/mattermost/interactions/default",
		auth: "plugin",
		handler: createMattermostInteractionHandler({
			client,
			botUserId,
			accountId: account.accountId,
			allowedSourceIps: effectiveInteractionSourceIps,
			trustedProxies: cfg.gateway?.trustedProxies,
			allowRealIpFallback: cfg.gateway?.allowRealIpFallback === true,
			handleInteraction: handleModelPickerInteraction,
			authorizeButtonClick: async ({ payload, post }) => {
				const channelInfo = await resolveChannelInfo(payload.channel_id);
				const isDirect = channelInfo?.type?.trim().toUpperCase() === "D";
				const allowTextCommands = core.channel.commands.shouldHandleTextCommands({
					cfg,
					surface: "mattermost"
				});
				const decision = authorizeMattermostCommandInvocation({
					account,
					cfg,
					senderId: payload.user_id,
					senderName: payload.user_name ?? "",
					channelId: payload.channel_id,
					channelInfo,
					storeAllowFrom: isDirect ? await readStoreAllowFromForDmPolicy({
						provider: "mattermost",
						accountId: account.accountId,
						dmPolicy: account.config.dmPolicy ?? "pairing",
						readStore: pairing.readStoreForDmPolicy
					}) : void 0,
					allowTextCommands,
					hasControlCommand: false
				});
				if (decision.ok) return { ok: true };
				return {
					ok: false,
					response: {
						update: {
							message: post.message ?? "",
							props: post.props ?? void 0
						},
						ephemeral_text: `OpenClaw ignored this action for ${decision.roomLabel}.`
					}
				};
			},
			resolveSessionKey: async ({ channelId, userId, post }) => {
				const channelInfo = await resolveChannelInfo(channelId);
				const kind = mapMattermostChannelTypeToChatType(channelInfo?.type);
				const teamId = channelInfo?.team_id ?? void 0;
				const route = core.channel.routing.resolveAgentRoute({
					cfg,
					channel: "mattermost",
					accountId: account.accountId,
					teamId,
					peer: {
						kind,
						id: kind === "direct" ? userId : channelId
					}
				});
				const replyToMode = resolveMattermostReplyToMode(account, kind);
				return resolveMattermostThreadSessionContext({
					baseSessionKey: route.sessionKey,
					kind,
					postId: post.id || void 0,
					replyToMode,
					threadRootId: post.root_id
				}).sessionKey;
			},
			dispatchButtonClick: async (opts) => {
				const channelInfo = await resolveChannelInfo(opts.channelId);
				const kind = mapMattermostChannelTypeToChatType(channelInfo?.type);
				const chatType = channelChatType(kind);
				const teamId = channelInfo?.team_id ?? void 0;
				const channelName = channelInfo?.name ?? void 0;
				const channelDisplay = channelInfo?.display_name ?? channelName ?? opts.channelId;
				const route = core.channel.routing.resolveAgentRoute({
					cfg,
					channel: "mattermost",
					accountId: account.accountId,
					teamId,
					peer: {
						kind,
						id: kind === "direct" ? opts.userId : opts.channelId
					}
				});
				const replyToMode = resolveMattermostReplyToMode(account, kind);
				const threadContext = resolveMattermostThreadSessionContext({
					baseSessionKey: route.sessionKey,
					kind,
					postId: opts.post.id || opts.postId,
					replyToMode,
					threadRootId: opts.post.root_id
				});
				const to = kind === "direct" ? `user:${opts.userId}` : `channel:${opts.channelId}`;
				const bodyText = `[Button click: user @${opts.userName} selected "${opts.actionName}"]`;
				const ctxPayload = core.channel.reply.finalizeInboundContext({
					Body: bodyText,
					BodyForAgent: bodyText,
					RawBody: bodyText,
					CommandBody: bodyText,
					From: kind === "direct" ? `mattermost:${opts.userId}` : kind === "group" ? `mattermost:group:${opts.channelId}` : `mattermost:channel:${opts.channelId}`,
					To: to,
					SessionKey: threadContext.sessionKey,
					ParentSessionKey: threadContext.parentSessionKey,
					AccountId: route.accountId,
					ChatType: chatType,
					ConversationLabel: `mattermost:${opts.userName}`,
					GroupSubject: kind !== "direct" ? channelDisplay : void 0,
					GroupChannel: channelName ? `#${channelName}` : void 0,
					GroupSpace: teamId,
					SenderName: opts.userName,
					SenderId: opts.userId,
					Provider: "mattermost",
					Surface: "mattermost",
					MessageSid: `interaction:${opts.postId}:${opts.actionId}`,
					ReplyToId: threadContext.effectiveReplyToId,
					MessageThreadId: threadContext.effectiveReplyToId,
					WasMentioned: true,
					CommandAuthorized: false,
					OriginatingChannel: "mattermost",
					OriginatingTo: to
				});
				const textLimit = core.channel.text.resolveTextChunkLimit(cfg, "mattermost", account.accountId, { fallbackLimit: account.textChunkLimit ?? 4e3 });
				const tableMode = core.channel.text.resolveMarkdownTableMode({
					cfg,
					channel: "mattermost",
					accountId: account.accountId
				});
				const { onModelSelected, typingCallbacks, ...replyPipeline } = createChannelReplyPipeline({
					cfg,
					agentId: route.agentId,
					channel: "mattermost",
					accountId: account.accountId,
					typing: {
						start: () => sendTypingIndicator(opts.channelId, threadContext.effectiveReplyToId),
						onStartError: (err) => {
							logTypingFailure({
								log: (message) => logger.debug?.(message),
								channel: "mattermost",
								target: opts.channelId,
								error: err
							});
						}
					}
				});
				const { dispatcher, replyOptions, markDispatchIdle } = core.channel.reply.createReplyDispatcherWithTyping({
					...replyPipeline,
					humanDelay: core.channel.reply.resolveHumanDelayConfig(cfg, route.agentId),
					deliver: async (payload) => {
						await deliverMattermostReplyPayload({
							core,
							cfg,
							payload,
							to,
							accountId: account.accountId,
							agentId: route.agentId,
							replyToId: resolveMattermostReplyRootId({
								threadRootId: threadContext.effectiveReplyToId,
								replyToId: payload.replyToId
							}),
							textLimit,
							tableMode,
							sendMessage: sendMessageMattermost
						});
						runtime.log?.(`delivered button-click reply to ${to}`);
					},
					onError: (err, info) => {
						runtime.error?.(`mattermost button-click ${info.kind} reply failed: ${String(err)}`);
					},
					onReplyStart: typingCallbacks?.onReplyStart
				});
				await core.channel.reply.dispatchReplyFromConfig({
					ctx: ctxPayload,
					cfg,
					dispatcher,
					replyOptions: {
						...replyOptions,
						disableBlockStreaming: typeof account.blockStreaming === "boolean" ? !account.blockStreaming : void 0,
						onModelSelected
					}
				});
				markDispatchIdle();
			},
			log: (msg) => runtime.log?.(msg)
		}),
		pluginId: "mattermost",
		source: "mattermost-interactions",
		accountId: account.accountId,
		log: (msg) => runtime.log?.(msg)
	});
	const logger = core.logging.getChildLogger({ module: "mattermost" });
	const logVerboseMessage = (message) => {
		if (!core.logging.shouldLogVerbose()) return;
		logger.debug?.(message);
	};
	const mediaMaxBytes = resolveChannelMediaMaxBytes({
		cfg,
		resolveChannelLimitMb: () => void 0,
		accountId: account.accountId
	}) ?? 8 * 1024 * 1024;
	const historyLimit = Math.max(0, cfg.messages?.groupChat?.historyLimit ?? 50);
	const channelHistories = /* @__PURE__ */ new Map();
	const defaultGroupPolicy = resolveDefaultGroupPolicy(cfg);
	const { groupPolicy, providerMissingFallbackApplied } = resolveAllowlistProviderRuntimeGroupPolicy({
		providerConfigPresent: cfg.channels?.mattermost !== void 0,
		groupPolicy: account.config.groupPolicy,
		defaultGroupPolicy
	});
	warnMissingProviderGroupPolicyFallbackOnce({
		providerMissingFallbackApplied,
		providerKey: "mattermost",
		accountId: account.accountId,
		log: (message) => logVerboseMessage(message)
	});
	const { resolveMattermostMedia, sendTypingIndicator, resolveChannelInfo, resolveUserInfo, updateModelPickerPost } = createMattermostMonitorResources({
		accountId: account.accountId,
		callbackUrl,
		client,
		logger: { debug: (message) => logger.debug?.(String(message)) },
		mediaMaxBytes,
		fetchRemoteMedia: (params) => core.channel.media.fetchRemoteMedia(params),
		saveMediaBuffer: (buffer, contentType, direction, maxBytes) => core.channel.media.saveMediaBuffer(Buffer.from(buffer), contentType, direction, maxBytes),
		mediaKindFromMime: (contentType) => core.media.mediaKindFromMime(contentType)
	});
	const runModelPickerCommand = async (params) => {
		const to = params.kind === "direct" ? `user:${params.senderId}` : `channel:${params.channelId}`;
		const fromLabel = params.kind === "direct" ? `Mattermost DM from ${params.senderName}` : `Mattermost message in ${params.roomLabel} from ${params.senderName}`;
		const ctxPayload = core.channel.reply.finalizeInboundContext({
			Body: params.commandText,
			BodyForAgent: params.commandText,
			RawBody: params.commandText,
			CommandBody: params.commandText,
			From: params.kind === "direct" ? `mattermost:${params.senderId}` : params.kind === "group" ? `mattermost:group:${params.channelId}` : `mattermost:channel:${params.channelId}`,
			To: to,
			SessionKey: params.sessionKey,
			ParentSessionKey: params.parentSessionKey,
			AccountId: params.route.accountId,
			ChatType: params.chatType,
			ConversationLabel: fromLabel,
			GroupSubject: params.kind !== "direct" ? params.channelDisplay || params.roomLabel : void 0,
			GroupChannel: params.channelName ? `#${params.channelName}` : void 0,
			GroupSpace: params.teamId,
			SenderName: params.senderName,
			SenderId: params.senderId,
			Provider: "mattermost",
			Surface: "mattermost",
			MessageSid: `interaction:${params.postId}:${Date.now()}`,
			ReplyToId: params.effectiveReplyToId,
			MessageThreadId: params.effectiveReplyToId,
			Timestamp: Date.now(),
			WasMentioned: true,
			CommandAuthorized: params.commandAuthorized,
			CommandSource: "native",
			OriginatingChannel: "mattermost",
			OriginatingTo: to
		});
		const tableMode = core.channel.text.resolveMarkdownTableMode({
			cfg,
			channel: "mattermost",
			accountId: account.accountId
		});
		const textLimit = core.channel.text.resolveTextChunkLimit(cfg, "mattermost", account.accountId, { fallbackLimit: account.textChunkLimit ?? 4e3 });
		const shouldDeliverReplies = params.deliverReplies === true;
		const { onModelSelected, typingCallbacks, ...replyPipeline } = createChannelReplyPipeline({
			cfg,
			agentId: params.route.agentId,
			channel: "mattermost",
			accountId: account.accountId,
			typing: shouldDeliverReplies ? {
				start: () => sendTypingIndicator(params.channelId, params.effectiveReplyToId),
				onStartError: (err) => {
					logTypingFailure({
						log: (message) => logger.debug?.(message),
						channel: "mattermost",
						target: params.channelId,
						error: err
					});
				}
			} : void 0
		});
		const capturedTexts = [];
		const { dispatcher, replyOptions, markDispatchIdle } = core.channel.reply.createReplyDispatcherWithTyping({
			...replyPipeline,
			deliver: async (payload) => {
				const trimmedPayload = {
					...payload,
					text: core.channel.text.convertMarkdownTables(payload.text ?? "", tableMode).trim()
				};
				if (!shouldDeliverReplies) {
					if (trimmedPayload.text) capturedTexts.push(trimmedPayload.text);
					return;
				}
				await deliverMattermostReplyPayload({
					core,
					cfg,
					payload: trimmedPayload,
					to,
					accountId: account.accountId,
					agentId: params.route.agentId,
					replyToId: resolveMattermostReplyRootId({
						threadRootId: params.effectiveReplyToId,
						replyToId: trimmedPayload.replyToId
					}),
					textLimit,
					tableMode: "off",
					sendMessage: sendMessageMattermost
				});
			},
			onError: (err, info) => {
				runtime.error?.(`mattermost model picker ${info.kind} reply failed: ${String(err)}`);
			},
			onReplyStart: typingCallbacks?.onReplyStart
		});
		await core.channel.reply.withReplyDispatcher({
			dispatcher,
			onSettled: () => {
				markDispatchIdle();
			},
			run: () => core.channel.reply.dispatchReplyFromConfig({
				ctx: ctxPayload,
				cfg,
				dispatcher,
				replyOptions: {
					...replyOptions,
					disableBlockStreaming: typeof account.blockStreaming === "boolean" ? !account.blockStreaming : void 0,
					onModelSelected
				}
			})
		});
		return capturedTexts.join("\n\n").trim();
	};
	async function handleModelPickerInteraction(params) {
		const pickerState = parseMattermostModelPickerContext(params.context);
		if (!pickerState) return null;
		if (pickerState.ownerUserId !== params.payload.user_id) return { ephemeral_text: "Only the person who opened this picker can use it." };
		const channelInfo = await resolveChannelInfo(params.payload.channel_id);
		const pickerCommandText = pickerState.action === "select" ? `/model ${pickerState.provider}/${pickerState.model}` : pickerState.action === "list" ? `/models ${pickerState.provider}` : "/models";
		const allowTextCommands = core.channel.commands.shouldHandleTextCommands({
			cfg,
			surface: "mattermost"
		});
		const hasControlCommand = core.channel.text.hasControlCommand(pickerCommandText, cfg);
		const dmPolicy = account.config.dmPolicy ?? "pairing";
		const storeAllowFrom = normalizeMattermostAllowList(await readStoreAllowFromForDmPolicy({
			provider: "mattermost",
			accountId: account.accountId,
			dmPolicy,
			readStore: pairing.readStoreForDmPolicy
		}));
		const auth = authorizeMattermostCommandInvocation({
			account,
			cfg,
			senderId: params.payload.user_id,
			senderName: params.userName,
			channelId: params.payload.channel_id,
			channelInfo,
			storeAllowFrom,
			allowTextCommands,
			hasControlCommand
		});
		if (!auth.ok) {
			if (auth.denyReason === "dm-pairing") {
				const { code } = await pairing.upsertPairingRequest({
					id: params.payload.user_id,
					meta: { name: params.userName }
				});
				return { ephemeral_text: core.channel.pairing.buildPairingReply({
					channel: "mattermost",
					idLine: `Your Mattermost user id: ${params.payload.user_id}`,
					code
				}) };
			}
			return { ephemeral_text: auth.denyReason === "unknown-channel" ? "Temporary error: unable to determine channel type. Please try again." : auth.denyReason === "dm-disabled" ? "This bot is not accepting direct messages." : auth.denyReason === "channels-disabled" ? "Model picker actions are disabled in channels." : auth.denyReason === "channel-no-allowlist" ? "Model picker actions are not configured for this channel." : "Unauthorized." };
		}
		const kind = auth.kind;
		const chatType = auth.chatType;
		const teamId = auth.channelInfo.team_id ?? params.payload.team_id ?? void 0;
		const channelName = auth.channelName || void 0;
		const channelDisplay = auth.channelDisplay || auth.channelName || params.payload.channel_id;
		const roomLabel = auth.roomLabel;
		const route = core.channel.routing.resolveAgentRoute({
			cfg,
			channel: "mattermost",
			accountId: account.accountId,
			teamId,
			peer: {
				kind,
				id: kind === "direct" ? params.payload.user_id : params.payload.channel_id
			}
		});
		const replyToMode = resolveMattermostReplyToMode(account, kind);
		const threadContext = resolveMattermostThreadSessionContext({
			baseSessionKey: route.sessionKey,
			kind,
			postId: params.post.id || params.payload.post_id,
			replyToMode,
			threadRootId: params.post.root_id
		});
		const modelSessionRoute = {
			agentId: route.agentId,
			sessionKey: threadContext.sessionKey
		};
		const data = await buildModelsProviderData(cfg, route.agentId);
		if (data.providers.length === 0) return await updateModelPickerPost({
			channelId: params.payload.channel_id,
			postId: params.payload.post_id,
			message: "No models available."
		});
		if (pickerState.action === "providers" || pickerState.action === "back") {
			const currentModel = resolveMattermostModelPickerCurrentModel({
				cfg,
				route: modelSessionRoute,
				data
			});
			const view = renderMattermostProviderPickerView({
				ownerUserId: pickerState.ownerUserId,
				data,
				currentModel
			});
			return await updateModelPickerPost({
				channelId: params.payload.channel_id,
				postId: params.payload.post_id,
				message: view.text,
				buttons: view.buttons
			});
		}
		if (pickerState.action === "list") {
			const currentModel = resolveMattermostModelPickerCurrentModel({
				cfg,
				route: modelSessionRoute,
				data
			});
			const view = renderMattermostModelsPickerView({
				ownerUserId: pickerState.ownerUserId,
				data,
				provider: pickerState.provider,
				page: pickerState.page,
				currentModel
			});
			return await updateModelPickerPost({
				channelId: params.payload.channel_id,
				postId: params.payload.post_id,
				message: view.text,
				buttons: view.buttons
			});
		}
		const targetModelRef = `${pickerState.provider}/${pickerState.model}`;
		if (!buildMattermostAllowedModelRefs(data).has(targetModelRef)) return { ephemeral_text: `That model is no longer available: ${targetModelRef}` };
		(async () => {
			try {
				await runModelPickerCommand({
					commandText: `/model ${targetModelRef}`,
					commandAuthorized: auth.commandAuthorized,
					route,
					sessionKey: threadContext.sessionKey,
					parentSessionKey: threadContext.parentSessionKey,
					channelId: params.payload.channel_id,
					senderId: params.payload.user_id,
					senderName: params.userName,
					kind,
					chatType,
					channelName,
					channelDisplay,
					roomLabel,
					teamId,
					postId: params.payload.post_id,
					effectiveReplyToId: threadContext.effectiveReplyToId,
					deliverReplies: true
				});
				const updatedModel = resolveMattermostModelPickerCurrentModel({
					cfg,
					route: modelSessionRoute,
					data,
					skipCache: true
				});
				const view = renderMattermostModelsPickerView({
					ownerUserId: pickerState.ownerUserId,
					data,
					provider: pickerState.provider,
					page: pickerState.page,
					currentModel: updatedModel
				});
				await updateModelPickerPost({
					channelId: params.payload.channel_id,
					postId: params.payload.post_id,
					message: view.text,
					buttons: view.buttons
				});
			} catch (err) {
				runtime.error?.(`mattermost model picker select failed: ${String(err)}`);
			}
		})();
		return {};
	}
	const handlePost = async (post, payload, messageIds) => {
		const channelId = post.channel_id ?? payload.data?.channel_id ?? payload.broadcast?.channel_id;
		if (!channelId) {
			logVerboseMessage("mattermost: drop post (missing channel id)");
			return;
		}
		const allMessageIds = messageIds?.length ? messageIds : post.id ? [post.id] : [];
		if (allMessageIds.length === 0) {
			logVerboseMessage("mattermost: drop post (missing message id)");
			return;
		}
		const dedupeEntries = allMessageIds.map((id) => recentInboundMessages.check(`${account.accountId}:${id}`));
		if (dedupeEntries.length > 0 && dedupeEntries.every(Boolean)) {
			logVerboseMessage(`mattermost: drop post (dedupe account=${account.accountId} ids=${allMessageIds.length})`);
			return;
		}
		const senderId = post.user_id ?? payload.broadcast?.user_id;
		if (!senderId) {
			logVerboseMessage("mattermost: drop post (missing sender id)");
			return;
		}
		if (senderId === botUserId) {
			logVerboseMessage(`mattermost: drop post (self sender=${senderId})`);
			return;
		}
		if (isSystemPost(post)) {
			logVerboseMessage(`mattermost: drop post (system post type=${post.type ?? "unknown"})`);
			return;
		}
		const channelInfo = await resolveChannelInfo(channelId);
		const kind = mapMattermostChannelTypeToChatType(payload.data?.channel_type ?? channelInfo?.type ?? void 0);
		const chatType = channelChatType(kind);
		const senderName = payload.data?.sender_name?.trim() || (await resolveUserInfo(senderId))?.username?.trim() || senderId;
		const rawText = post.message?.trim() || "";
		const dmPolicy = account.config.dmPolicy ?? "pairing";
		const normalizedAllowFrom = normalizeMattermostAllowList(account.config.allowFrom ?? []);
		const normalizedGroupAllowFrom = normalizeMattermostAllowList(account.config.groupAllowFrom ?? []);
		const storeAllowFrom = normalizeMattermostAllowList(await readStoreAllowFromForDmPolicy({
			provider: "mattermost",
			accountId: account.accountId,
			dmPolicy,
			readStore: pairing.readStoreForDmPolicy
		}));
		const accessDecision = resolveDmGroupAccessWithLists({
			isGroup: kind !== "direct",
			dmPolicy,
			groupPolicy,
			allowFrom: normalizedAllowFrom,
			groupAllowFrom: normalizedGroupAllowFrom,
			storeAllowFrom,
			isSenderAllowed: (allowFrom) => isMattermostSenderAllowed({
				senderId,
				senderName,
				allowFrom,
				allowNameMatching
			})
		});
		const effectiveAllowFrom = accessDecision.effectiveAllowFrom;
		const effectiveGroupAllowFrom = accessDecision.effectiveGroupAllowFrom;
		const allowTextCommands = core.channel.commands.shouldHandleTextCommands({
			cfg,
			surface: "mattermost"
		});
		const hasControlCommand = core.channel.text.hasControlCommand(rawText, cfg);
		const isControlCommand = allowTextCommands && hasControlCommand;
		const useAccessGroups = cfg.commands?.useAccessGroups !== false;
		const commandDmAllowFrom = kind === "direct" ? effectiveAllowFrom : normalizedAllowFrom;
		const senderAllowedForCommands = isMattermostSenderAllowed({
			senderId,
			senderName,
			allowFrom: commandDmAllowFrom,
			allowNameMatching
		});
		const groupAllowedForCommands = isMattermostSenderAllowed({
			senderId,
			senderName,
			allowFrom: effectiveGroupAllowFrom,
			allowNameMatching
		});
		const commandGate = resolveControlCommandGate({
			useAccessGroups,
			authorizers: [{
				configured: commandDmAllowFrom.length > 0,
				allowed: senderAllowedForCommands
			}, {
				configured: effectiveGroupAllowFrom.length > 0,
				allowed: groupAllowedForCommands
			}],
			allowTextCommands,
			hasControlCommand
		});
		const commandAuthorized = commandGate.commandAuthorized;
		if (accessDecision.decision !== "allow") {
			if (kind === "direct") {
				if (accessDecision.reasonCode === DM_GROUP_ACCESS_REASON.DM_POLICY_DISABLED) {
					logVerboseMessage(`mattermost: drop dm (dmPolicy=disabled sender=${senderId})`);
					return;
				}
				if (accessDecision.decision === "pairing") {
					const { code, created } = await pairing.upsertPairingRequest({
						id: senderId,
						meta: { name: senderName }
					});
					logVerboseMessage(`mattermost: pairing request sender=${senderId} created=${created}`);
					if (created) try {
						await sendMessageMattermost(`user:${senderId}`, core.channel.pairing.buildPairingReply({
							channel: "mattermost",
							idLine: `Your Mattermost user id: ${senderId}`,
							code
						}), {
							cfg,
							accountId: account.accountId
						});
						opts.statusSink?.({ lastOutboundAt: Date.now() });
					} catch (err) {
						logVerboseMessage(`mattermost: pairing reply failed for ${senderId}: ${String(err)}`);
					}
					return;
				}
				logVerboseMessage(`mattermost: drop dm sender=${senderId} (dmPolicy=${dmPolicy})`);
				return;
			}
			if (accessDecision.reasonCode === DM_GROUP_ACCESS_REASON.GROUP_POLICY_DISABLED) {
				logVerboseMessage("mattermost: drop group message (groupPolicy=disabled)");
				return;
			}
			if (accessDecision.reasonCode === DM_GROUP_ACCESS_REASON.GROUP_POLICY_EMPTY_ALLOWLIST) {
				logVerboseMessage("mattermost: drop group message (no group allowlist)");
				return;
			}
			if (accessDecision.reasonCode === DM_GROUP_ACCESS_REASON.GROUP_POLICY_NOT_ALLOWLISTED) {
				logVerboseMessage(`mattermost: drop group sender=${senderId} (not in groupAllowFrom)`);
				return;
			}
			logVerboseMessage(`mattermost: drop group message (groupPolicy=${groupPolicy} reason=${accessDecision.reason})`);
			return;
		}
		if (kind !== "direct" && commandGate.shouldBlock) {
			logInboundDrop({
				log: logVerboseMessage,
				channel: "mattermost",
				reason: "control command (unauthorized)",
				target: senderId
			});
			return;
		}
		const teamId = payload.data?.team_id ?? channelInfo?.team_id ?? void 0;
		const channelName = payload.data?.channel_name ?? channelInfo?.name ?? "";
		const channelDisplay = payload.data?.channel_display_name ?? channelInfo?.display_name ?? channelName;
		const roomLabel = channelName ? `#${channelName}` : channelDisplay || `#${channelId}`;
		const route = core.channel.routing.resolveAgentRoute({
			cfg,
			channel: "mattermost",
			accountId: account.accountId,
			teamId,
			peer: {
				kind,
				id: kind === "direct" ? senderId : channelId
			}
		});
		const baseSessionKey = route.sessionKey;
		const threadRootId = post.root_id?.trim() || void 0;
		const replyToMode = resolveMattermostReplyToMode(account, kind);
		const { effectiveReplyToId, sessionKey, parentSessionKey } = resolveMattermostThreadSessionContext({
			baseSessionKey,
			kind,
			postId: post.id,
			replyToMode,
			threadRootId
		});
		const historyKey = kind === "direct" ? null : sessionKey;
		const mentionRegexes = core.channel.mentions.buildMentionRegexes(cfg, route.agentId);
		const wasMentioned = kind !== "direct" && ((botUsername ? rawText.toLowerCase().includes(`@${botUsername.toLowerCase()}`) : false) || core.channel.mentions.matchesMentionPatterns(rawText, mentionRegexes));
		const pendingBody = rawText || (post.file_ids?.length ? `[Mattermost ${post.file_ids.length === 1 ? "file" : "files"}]` : "");
		const pendingSender = senderName;
		const recordPendingHistory = () => {
			const trimmed = pendingBody.trim();
			recordPendingHistoryEntryIfEnabled({
				historyMap: channelHistories,
				limit: historyLimit,
				historyKey: historyKey ?? "",
				entry: historyKey && trimmed ? {
					sender: pendingSender,
					body: trimmed,
					timestamp: typeof post.create_at === "number" ? post.create_at : void 0,
					messageId: post.id ?? void 0
				} : null
			});
		};
		const oncharEnabled = account.chatmode === "onchar" && kind !== "direct";
		const oncharPrefixes = oncharEnabled ? resolveOncharPrefixes(account.oncharPrefixes) : [];
		const oncharResult = oncharEnabled ? stripOncharPrefix(rawText, oncharPrefixes) : {
			triggered: false,
			stripped: rawText
		};
		const oncharTriggered = oncharResult.triggered;
		const canDetectMention = Boolean(botUsername) || mentionRegexes.length > 0;
		const mentionDecision = evaluateMattermostMentionGate({
			kind,
			cfg,
			accountId: account.accountId,
			channelId,
			threadRootId,
			requireMentionOverride: account.requireMention,
			resolveRequireMention: core.channel.groups.resolveRequireMention,
			wasMentioned,
			isControlCommand,
			commandAuthorized,
			oncharEnabled,
			oncharTriggered,
			canDetectMention
		});
		const { shouldRequireMention, shouldBypassMention } = mentionDecision;
		if (mentionDecision.dropReason === "onchar-not-triggered") {
			logVerboseMessage(`mattermost: drop group message (onchar not triggered channel=${channelId} sender=${senderId})`);
			recordPendingHistory();
			return;
		}
		if (mentionDecision.dropReason === "missing-mention") {
			logVerboseMessage(`mattermost: drop group message (missing mention channel=${channelId} sender=${senderId} requireMention=${shouldRequireMention} bypass=${shouldBypassMention} canDetectMention=${canDetectMention})`);
			recordPendingHistory();
			return;
		}
		const mediaList = await resolveMattermostMedia(post.file_ids);
		const mediaPlaceholder = buildMattermostAttachmentPlaceholder(mediaList);
		const bodyText = normalizeMention([oncharTriggered ? oncharResult.stripped : rawText, mediaPlaceholder].filter(Boolean).join("\n").trim(), botUsername);
		if (!bodyText) {
			logVerboseMessage(`mattermost: drop group message (empty body after normalization channel=${channelId} sender=${senderId})`);
			return;
		}
		core.channel.activity.record({
			channel: "mattermost",
			accountId: account.accountId,
			direction: "inbound"
		});
		const fromLabel = formatInboundFromLabel({
			isGroup: kind !== "direct",
			groupLabel: channelDisplay || roomLabel,
			groupId: channelId,
			groupFallback: roomLabel || "Channel",
			directLabel: senderName,
			directId: senderId
		});
		const preview = bodyText.replace(/\s+/g, " ").slice(0, 160);
		const inboundLabel = kind === "direct" ? `Mattermost DM from ${senderName}` : `Mattermost message in ${roomLabel} from ${senderName}`;
		core.system.enqueueSystemEvent(`${inboundLabel}: ${preview}`, {
			sessionKey,
			contextKey: `mattermost:message:${channelId}:${post.id ?? "unknown"}`
		});
		const textWithId = `${bodyText}\n[mattermost message id: ${post.id ?? "unknown"} channel: ${channelId}]`;
		let combinedBody = core.channel.reply.formatInboundEnvelope({
			channel: "Mattermost",
			from: fromLabel,
			timestamp: typeof post.create_at === "number" ? post.create_at : void 0,
			body: textWithId,
			chatType,
			sender: {
				name: senderName,
				id: senderId
			}
		});
		if (historyKey) combinedBody = buildPendingHistoryContextFromMap({
			historyMap: channelHistories,
			historyKey,
			limit: historyLimit,
			currentMessage: combinedBody,
			formatEntry: (entry) => core.channel.reply.formatInboundEnvelope({
				channel: "Mattermost",
				from: fromLabel,
				timestamp: entry.timestamp,
				body: `${entry.body}${entry.messageId ? ` [id:${entry.messageId} channel:${channelId}]` : ""}`,
				chatType,
				senderLabel: entry.sender
			})
		});
		const to = kind === "direct" ? `user:${senderId}` : `channel:${channelId}`;
		const mediaPayload = buildAgentMediaPayload(mediaList);
		const commandBody = rawText.trim();
		const inboundHistory = historyKey && historyLimit > 0 ? (channelHistories.get(historyKey) ?? []).map((entry) => ({
			sender: entry.sender,
			body: entry.body,
			timestamp: entry.timestamp
		})) : void 0;
		const ctxPayload = core.channel.reply.finalizeInboundContext({
			Body: combinedBody,
			BodyForAgent: bodyText,
			InboundHistory: inboundHistory,
			RawBody: bodyText,
			CommandBody: commandBody,
			BodyForCommands: commandBody,
			From: kind === "direct" ? `mattermost:${senderId}` : kind === "group" ? `mattermost:group:${channelId}` : `mattermost:channel:${channelId}`,
			To: to,
			SessionKey: sessionKey,
			ParentSessionKey: parentSessionKey,
			AccountId: route.accountId,
			ChatType: chatType,
			ConversationLabel: fromLabel,
			GroupSubject: kind !== "direct" ? channelDisplay || roomLabel : void 0,
			GroupChannel: channelName ? `#${channelName}` : void 0,
			GroupSpace: teamId,
			SenderName: senderName,
			SenderId: senderId,
			Provider: "mattermost",
			Surface: "mattermost",
			MessageSid: post.id ?? void 0,
			MessageSids: allMessageIds.length > 1 ? allMessageIds : void 0,
			MessageSidFirst: allMessageIds.length > 1 ? allMessageIds[0] : void 0,
			MessageSidLast: allMessageIds.length > 1 ? allMessageIds[allMessageIds.length - 1] : void 0,
			ReplyToId: effectiveReplyToId,
			MessageThreadId: effectiveReplyToId,
			Timestamp: typeof post.create_at === "number" ? post.create_at : void 0,
			WasMentioned: kind !== "direct" ? mentionDecision.effectiveWasMentioned : void 0,
			CommandAuthorized: commandAuthorized,
			OriginatingChannel: "mattermost",
			OriginatingTo: to,
			...mediaPayload
		});
		if (kind === "direct") {
			const sessionCfg = cfg.session;
			const storePath = core.channel.session.resolveStorePath(sessionCfg?.store, { agentId: route.agentId });
			await core.channel.session.updateLastRoute({
				storePath,
				sessionKey: route.mainSessionKey,
				deliveryContext: {
					channel: "mattermost",
					to,
					accountId: route.accountId
				}
			});
		}
		const previewLine = bodyText.slice(0, 200).replace(/\n/g, "\\n");
		logVerboseMessage(`mattermost inbound: from=${ctxPayload.From} len=${bodyText.length} preview="${previewLine}"`);
		const textLimit = core.channel.text.resolveTextChunkLimit(cfg, "mattermost", account.accountId, { fallbackLimit: account.textChunkLimit ?? 4e3 });
		const tableMode = core.channel.text.resolveMarkdownTableMode({
			cfg,
			channel: "mattermost",
			accountId: account.accountId
		});
		const { onModelSelected, typingCallbacks, ...replyPipeline } = createChannelReplyPipeline({
			cfg,
			agentId: route.agentId,
			channel: "mattermost",
			accountId: account.accountId,
			typing: {
				start: () => sendTypingIndicator(channelId, effectiveReplyToId),
				onStartError: (err) => {
					logTypingFailure({
						log: (message) => logger.debug?.(message),
						channel: "mattermost",
						target: channelId,
						error: err
					});
				}
			}
		});
		const { dispatcher, replyOptions, markDispatchIdle } = core.channel.reply.createReplyDispatcherWithTyping({
			...replyPipeline,
			humanDelay: core.channel.reply.resolveHumanDelayConfig(cfg, route.agentId),
			typingCallbacks,
			deliver: async (payload) => {
				await deliverMattermostReplyPayload({
					core,
					cfg,
					payload,
					to,
					accountId: account.accountId,
					agentId: route.agentId,
					replyToId: resolveMattermostReplyRootId({
						threadRootId: effectiveReplyToId,
						replyToId: payload.replyToId
					}),
					textLimit,
					tableMode,
					sendMessage: sendMessageMattermost
				});
				runtime.log?.(`delivered reply to ${to}`);
			},
			onError: (err, info) => {
				runtime.error?.(`mattermost ${info.kind} reply failed: ${String(err)}`);
			}
		});
		await core.channel.reply.withReplyDispatcher({
			dispatcher,
			onSettled: () => {
				markDispatchIdle();
			},
			run: () => core.channel.reply.dispatchReplyFromConfig({
				ctx: ctxPayload,
				cfg,
				dispatcher,
				replyOptions: {
					...replyOptions,
					disableBlockStreaming: typeof account.blockStreaming === "boolean" ? !account.blockStreaming : void 0,
					onModelSelected
				}
			})
		});
		if (historyKey) clearHistoryEntriesIfEnabled({
			historyMap: channelHistories,
			historyKey,
			limit: historyLimit
		});
	};
	const handleReactionEvent = async (payload) => {
		const reactionData = payload.data?.reaction;
		if (!reactionData) return;
		let reaction = null;
		if (typeof reactionData === "string") try {
			reaction = JSON.parse(reactionData);
		} catch {
			return;
		}
		else if (typeof reactionData === "object") reaction = reactionData;
		if (!reaction) return;
		const userId = reaction.user_id?.trim();
		const postId = reaction.post_id?.trim();
		const emojiName = reaction.emoji_name?.trim();
		if (!userId || !postId || !emojiName) return;
		if (userId === botUserId) return;
		const action = payload.event === "reaction_removed" ? "removed" : "added";
		const senderName = (await resolveUserInfo(userId))?.username?.trim() || userId;
		const channelId = resolveMattermostReactionChannelId(payload);
		if (!channelId) {
			logVerboseMessage(`mattermost: drop reaction (no channel_id in broadcast, cannot enforce policy)`);
			return;
		}
		const channelInfo = await resolveChannelInfo(channelId);
		if (!channelInfo?.type) {
			logVerboseMessage(`mattermost: drop reaction (cannot resolve channel type for ${channelId})`);
			return;
		}
		const kind = mapMattermostChannelTypeToChatType(channelInfo.type);
		const dmPolicy = account.config.dmPolicy ?? "pairing";
		const storeAllowFrom = normalizeMattermostAllowList(await readStoreAllowFromForDmPolicy({
			provider: "mattermost",
			accountId: account.accountId,
			dmPolicy,
			readStore: pairing.readStoreForDmPolicy
		}));
		const reactionAccess = resolveDmGroupAccessWithLists({
			isGroup: kind !== "direct",
			dmPolicy,
			groupPolicy,
			allowFrom: normalizeMattermostAllowList(account.config.allowFrom ?? []),
			groupAllowFrom: normalizeMattermostAllowList(account.config.groupAllowFrom ?? []),
			storeAllowFrom,
			isSenderAllowed: (allowFrom) => isMattermostSenderAllowed({
				senderId: userId,
				senderName,
				allowFrom,
				allowNameMatching
			})
		});
		if (reactionAccess.decision !== "allow") {
			if (kind === "direct") logVerboseMessage(`mattermost: drop reaction (dmPolicy=${dmPolicy} sender=${userId} reason=${reactionAccess.reason})`);
			else logVerboseMessage(`mattermost: drop reaction (groupPolicy=${groupPolicy} sender=${userId} reason=${reactionAccess.reason} channel=${channelId})`);
			return;
		}
		const teamId = channelInfo?.team_id ?? void 0;
		const sessionKey = core.channel.routing.resolveAgentRoute({
			cfg,
			channel: "mattermost",
			accountId: account.accountId,
			teamId,
			peer: {
				kind,
				id: kind === "direct" ? userId : channelId
			}
		}).sessionKey;
		const eventText = `Mattermost reaction ${action}: :${emojiName}: by @${senderName} on post ${postId} in channel ${channelId}`;
		core.system.enqueueSystemEvent(eventText, {
			sessionKey,
			contextKey: `mattermost:reaction:${postId}:${emojiName}:${userId}:${action}`
		});
		logVerboseMessage(`mattermost reaction: ${action} :${emojiName}: by ${senderName} on ${postId}`);
	};
	const inboundDebounceMs = core.channel.debounce.resolveInboundDebounceMs({
		cfg,
		channel: "mattermost"
	});
	const debouncer = core.channel.debounce.createInboundDebouncer({
		debounceMs: inboundDebounceMs,
		buildKey: (entry) => {
			const channelId = entry.post.channel_id ?? entry.payload.data?.channel_id ?? entry.payload.broadcast?.channel_id;
			if (!channelId) return null;
			const threadId = entry.post.root_id?.trim();
			const threadKey = threadId ? `thread:${threadId}` : "channel";
			return `mattermost:${account.accountId}:${channelId}:${threadKey}`;
		},
		shouldDebounce: (entry) => {
			if (entry.post.file_ids && entry.post.file_ids.length > 0) return false;
			const text = entry.post.message?.trim() ?? "";
			if (!text) return false;
			return !core.channel.text.hasControlCommand(text, cfg);
		},
		onFlush: async (entries) => {
			const last = entries.at(-1);
			if (!last) return;
			if (entries.length === 1) {
				await handlePost(last.post, last.payload);
				return;
			}
			const combinedText = entries.map((entry) => entry.post.message?.trim() ?? "").filter(Boolean).join("\n");
			const mergedPost = {
				...last.post,
				message: combinedText,
				file_ids: []
			};
			const ids = entries.map((entry) => entry.post.id).filter(Boolean);
			await handlePost(mergedPost, last.payload, ids.length > 0 ? ids : void 0);
		},
		onError: (err) => {
			runtime.error?.(`mattermost debounce flush failed: ${String(err)}`);
		}
	});
	const wsUrl = buildMattermostWsUrl(baseUrl);
	let seq = 1;
	const connectOnce = createMattermostConnectOnce({
		wsUrl,
		botToken,
		abortSignal: opts.abortSignal,
		statusSink: opts.statusSink,
		runtime,
		webSocketFactory: opts.webSocketFactory,
		nextSeq: () => seq++,
		getBotUpdateAt: async () => {
			return (await fetchMattermostMe(client)).update_at ?? 0;
		},
		onPosted: async (post, payload) => {
			await debouncer.enqueue({
				post,
				payload
			});
		},
		onReaction: async (payload) => {
			await handleReactionEvent(payload);
		}
	});
	let slashShutdownCleanup = null;
	if (slashEnabled) {
		const runAbortCleanup = () => {
			if (slashShutdownCleanup) return;
			const commands = getSlashCommandState(account.accountId)?.registeredCommands ?? [];
			deactivateSlashCommands(account.accountId);
			slashShutdownCleanup = cleanupSlashCommands({
				client,
				commands,
				log: (msg) => runtime.log?.(msg)
			}).catch((err) => {
				runtime.error?.(`mattermost: slash cleanup failed: ${String(err)}`);
			});
		};
		if (opts.abortSignal?.aborted) runAbortCleanup();
		else opts.abortSignal?.addEventListener("abort", runAbortCleanup, { once: true });
	}
	try {
		await runWithReconnect(connectOnce, {
			abortSignal: opts.abortSignal,
			jitterRatio: .2,
			onError: (err) => {
				runtime.error?.(`mattermost connection failed: ${String(err)}`);
				opts.statusSink?.({
					lastError: String(err),
					connected: false
				});
			},
			onReconnect: (delayMs) => {
				runtime.log?.(`mattermost reconnecting in ${Math.round(delayMs / 1e3)}s`);
			}
		});
	} finally {
		unregisterInteractions?.();
	}
	if (slashShutdownCleanup) await slashShutdownCleanup;
}
//#endregion
//#region extensions/mattermost/src/mattermost/probe.ts
async function probeMattermost(baseUrl, botToken, timeoutMs = 2500, allowPrivateNetwork = false) {
	const normalized = normalizeMattermostBaseUrl(baseUrl);
	if (!normalized) return {
		ok: false,
		error: "baseUrl missing"
	};
	const url = `${normalized}/api/v4/users/me`;
	const start = Date.now();
	const controller = timeoutMs > 0 ? new AbortController() : void 0;
	let timer = null;
	if (controller) timer = setTimeout(() => controller.abort(), timeoutMs);
	try {
		const { response: res, release } = await fetchWithSsrFGuard({
			url,
			init: {
				headers: { Authorization: `Bearer ${botToken}` },
				signal: controller?.signal
			},
			auditContext: "mattermost-probe",
			policy: ssrfPolicyFromPrivateNetworkOptIn(allowPrivateNetwork)
		});
		try {
			const elapsedMs = Date.now() - start;
			if (!res.ok) {
				const detail = await readMattermostError(res);
				return {
					ok: false,
					status: res.status,
					error: detail || res.statusText,
					elapsedMs
				};
			}
			const bot = await res.json();
			return {
				ok: true,
				status: res.status,
				elapsedMs,
				bot
			};
		} finally {
			await release();
		}
	} catch (err) {
		return {
			ok: false,
			status: null,
			error: err instanceof Error ? err.message : String(err),
			elapsedMs: Date.now() - start
		};
	} finally {
		if (timer) clearTimeout(timer);
	}
}
//#endregion
//#region extensions/mattermost/src/mattermost/reactions.ts
const BOT_USER_CACHE_TTL_MS = 10 * 6e4;
const botUserIdCache = /* @__PURE__ */ new Map();
async function resolveBotUserId(client, cacheKey) {
	const cached = botUserIdCache.get(cacheKey);
	if (cached && cached.expiresAt > Date.now()) return cached.userId;
	const userId = (await fetchMattermostMe(client))?.id?.trim();
	if (!userId) return null;
	botUserIdCache.set(cacheKey, {
		userId,
		expiresAt: Date.now() + BOT_USER_CACHE_TTL_MS
	});
	return userId;
}
async function addMattermostReaction(params) {
	return runMattermostReaction(params, {
		action: "add",
		mutation: createReaction
	});
}
async function removeMattermostReaction(params) {
	return runMattermostReaction(params, {
		action: "remove",
		mutation: deleteReaction
	});
}
async function runMattermostReaction(params, options) {
	const resolved = resolveMattermostAccount({
		cfg: params.cfg,
		accountId: params.accountId
	});
	const baseUrl = resolved.baseUrl?.trim();
	const botToken = resolved.botToken?.trim();
	if (!baseUrl || !botToken) return {
		ok: false,
		error: "Mattermost botToken/baseUrl missing."
	};
	const client = createMattermostClient({
		baseUrl,
		botToken,
		fetchImpl: params.fetchImpl,
		allowPrivateNetwork: isPrivateNetworkOptInEnabled(resolved.config)
	});
	const userId = await resolveBotUserId(client, `${baseUrl}:${botToken}`);
	if (!userId) return {
		ok: false,
		error: "Mattermost reactions failed: could not resolve bot user id."
	};
	try {
		await options.mutation(client, {
			userId,
			postId: params.postId,
			emojiName: params.emojiName
		});
	} catch (err) {
		return {
			ok: false,
			error: `Mattermost ${options.action} reaction failed: ${String(err)}`
		};
	}
	return { ok: true };
}
async function createReaction(client, params) {
	await client.request("/reactions", {
		method: "POST",
		body: JSON.stringify({
			user_id: params.userId,
			post_id: params.postId,
			emoji_name: params.emojiName
		})
	});
}
async function deleteReaction(client, params) {
	const emoji = encodeURIComponent(params.emojiName);
	await client.request(`/users/${params.userId}/posts/${params.postId}/reactions/${emoji}`, { method: "DELETE" });
}
//#endregion
//#region extensions/mattermost/src/normalize.ts
function normalizeMattermostMessagingTarget(raw) {
	const trimmed = raw.trim();
	if (!trimmed) return;
	const lower = trimmed.toLowerCase();
	if (lower.startsWith("channel:")) {
		const id = trimmed.slice(8).trim();
		return id ? `channel:${id}` : void 0;
	}
	if (lower.startsWith("group:")) {
		const id = trimmed.slice(6).trim();
		return id ? `channel:${id}` : void 0;
	}
	if (lower.startsWith("user:")) {
		const id = trimmed.slice(5).trim();
		return id ? `user:${id}` : void 0;
	}
	if (lower.startsWith("mattermost:")) {
		const id = trimmed.slice(11).trim();
		return id ? `user:${id}` : void 0;
	}
	if (trimmed.startsWith("@")) {
		const id = trimmed.slice(1).trim();
		return id ? `@${id}` : void 0;
	}
	if (trimmed.startsWith("#")) return;
}
function looksLikeMattermostTargetId(raw, normalized) {
	const trimmed = raw.trim();
	if (!trimmed) return false;
	if (/^(user|channel|group|mattermost):/i.test(trimmed)) return true;
	if (trimmed.startsWith("@")) return true;
	return /^[a-z0-9]{26}$/i.test(trimmed) || /^[a-z0-9]{26}__[a-z0-9]{26}$/i.test(trimmed);
}
//#endregion
//#region extensions/mattermost/src/session-route.ts
function resolveMattermostOutboundSessionRoute(params) {
	let trimmed = stripChannelTargetPrefix(params.target, "mattermost");
	if (!trimmed) return null;
	const lower = trimmed.toLowerCase();
	const resolvedKind = params.resolvedTarget?.kind;
	const isUser = resolvedKind === "user" || resolvedKind !== "channel" && resolvedKind !== "group" && (lower.startsWith("user:") || trimmed.startsWith("@"));
	if (trimmed.startsWith("@")) trimmed = trimmed.slice(1).trim();
	const rawId = stripTargetKindPrefix(trimmed);
	if (!rawId) return null;
	const baseRoute = buildChannelOutboundSessionRoute({
		cfg: params.cfg,
		agentId: params.agentId,
		channel: "mattermost",
		accountId: params.accountId,
		peer: {
			kind: isUser ? "direct" : "channel",
			id: rawId
		},
		chatType: isUser ? "direct" : "channel",
		from: isUser ? `mattermost:${rawId}` : `mattermost:channel:${rawId}`,
		to: isUser ? `user:${rawId}` : `channel:${rawId}`
	});
	const threadId = normalizeOutboundThreadId(params.replyToId ?? params.threadId);
	const threadKeys = resolveThreadSessionKeys$1({
		baseSessionKey: baseRoute.baseSessionKey,
		threadId
	});
	return {
		...baseRoute,
		sessionKey: threadKeys.sessionKey,
		...threadId !== void 0 ? { threadId } : {}
	};
}
//#endregion
//#region extensions/mattermost/src/setup-core.ts
const channel$1 = "mattermost";
function isMattermostConfigured(account) {
	return (Boolean(account.botToken?.trim()) || hasConfiguredSecretInput(account.config.botToken)) && Boolean(account.baseUrl);
}
function resolveMattermostAccountWithSecrets(cfg, accountId) {
	return resolveMattermostAccount({
		cfg,
		accountId,
		allowUnresolvedSecretRef: true
	});
}
const mattermostSetupAdapter = {
	resolveAccountId: ({ accountId }) => normalizeAccountId(accountId),
	applyAccountName: ({ cfg, accountId, name }) => applyAccountNameToChannelSection({
		cfg,
		channelKey: channel$1,
		accountId,
		name
	}),
	validateInput: createSetupInputPresenceValidator({
		defaultAccountOnlyEnvError: "Mattermost env vars can only be used for the default account.",
		whenNotUseEnv: [{
			someOf: ["botToken", "token"],
			message: "Mattermost requires --bot-token and --http-url (or --use-env)."
		}, {
			someOf: ["httpUrl"],
			message: "Mattermost requires --bot-token and --http-url (or --use-env)."
		}],
		validate: ({ accountId, input }) => {
			const token = input.botToken ?? input.token;
			const baseUrl = normalizeMattermostBaseUrl(input.httpUrl);
			if (!input.useEnv && (!token || !baseUrl)) return "Mattermost requires --bot-token and --http-url (or --use-env).";
			if (input.httpUrl && !baseUrl) return "Mattermost --http-url must include a valid base URL.";
			return null;
		}
	}),
	applyAccountConfig: ({ cfg, accountId, input }) => {
		const token = input.botToken ?? input.token;
		const baseUrl = normalizeMattermostBaseUrl(input.httpUrl);
		const namedConfig = applyAccountNameToChannelSection({
			cfg,
			channelKey: channel$1,
			accountId,
			name: input.name
		});
		return applySetupAccountConfigPatch({
			cfg: accountId !== "default" ? migrateBaseNameToDefaultAccount({
				cfg: namedConfig,
				channelKey: channel$1
			}) : namedConfig,
			channelKey: channel$1,
			accountId,
			patch: input.useEnv ? {} : {
				...token ? { botToken: token } : {},
				...baseUrl ? { baseUrl } : {}
			}
		});
	}
};
//#endregion
//#region extensions/mattermost/src/setup-surface.ts
const channel = "mattermost";
const mattermostSetupWizard = {
	channel,
	status: createStandardChannelSetupStatus({
		channelLabel: "Mattermost",
		configuredLabel: "configured",
		unconfiguredLabel: "needs token + url",
		configuredHint: "configured",
		unconfiguredHint: "needs setup",
		configuredScore: 2,
		unconfiguredScore: 1,
		resolveConfigured: ({ cfg, accountId }) => isMattermostConfigured(resolveMattermostAccountWithSecrets(cfg, accountId ?? "default"))
	}),
	introNote: {
		title: "Mattermost bot token",
		lines: [
			"1) Mattermost System Console -> Integrations -> Bot Accounts",
			"2) Create a bot + copy its token",
			"3) Use your server base URL (e.g., https://chat.example.com)",
			"Tip: the bot must be a member of any channel you want it to monitor.",
			`Docs: ${formatDocsLink("/mattermost", "mattermost")}`
		],
		shouldShow: ({ cfg, accountId }) => !isMattermostConfigured(resolveMattermostAccountWithSecrets(cfg, accountId))
	},
	envShortcut: {
		prompt: "MATTERMOST_BOT_TOKEN + MATTERMOST_URL detected. Use env vars?",
		preferredEnvVar: "MATTERMOST_BOT_TOKEN",
		isAvailable: ({ cfg, accountId }) => {
			if (accountId !== "default") return false;
			const resolvedAccount = resolveMattermostAccountWithSecrets(cfg, accountId);
			const hasConfigValues = hasConfiguredSecretInput(resolvedAccount.config.botToken) || Boolean(resolvedAccount.config.baseUrl?.trim());
			return Boolean(process.env.MATTERMOST_BOT_TOKEN?.trim() && process.env.MATTERMOST_URL?.trim() && !hasConfigValues);
		},
		apply: ({ cfg, accountId }) => applySetupAccountConfigPatch({
			cfg,
			channelKey: channel,
			accountId,
			patch: {}
		})
	},
	credentials: [{
		inputKey: "botToken",
		providerHint: channel,
		credentialLabel: "bot token",
		preferredEnvVar: "MATTERMOST_BOT_TOKEN",
		envPrompt: "MATTERMOST_BOT_TOKEN + MATTERMOST_URL detected. Use env vars?",
		keepPrompt: "Mattermost bot token already configured. Keep it?",
		inputPrompt: "Enter Mattermost bot token",
		inspect: ({ cfg, accountId }) => {
			const resolvedAccount = resolveMattermostAccountWithSecrets(cfg, accountId);
			return {
				accountConfigured: isMattermostConfigured(resolvedAccount),
				hasConfiguredValue: hasConfiguredSecretInput(resolvedAccount.config.botToken)
			};
		}
	}],
	textInputs: [{
		inputKey: "httpUrl",
		message: "Enter Mattermost base URL",
		confirmCurrentValue: false,
		currentValue: ({ cfg, accountId }) => resolveMattermostAccountWithSecrets(cfg, accountId).baseUrl ?? process.env.MATTERMOST_URL?.trim(),
		initialValue: ({ cfg, accountId }) => resolveMattermostAccountWithSecrets(cfg, accountId).baseUrl ?? process.env.MATTERMOST_URL?.trim(),
		shouldPrompt: ({ cfg, accountId, credentialValues, currentValue }) => {
			const resolvedAccount = resolveMattermostAccountWithSecrets(cfg, accountId);
			const tokenConfigured = Boolean(resolvedAccount.botToken?.trim()) || hasConfiguredSecretInput(resolvedAccount.config.botToken);
			return Boolean(credentialValues.botToken) || !tokenConfigured || !currentValue;
		},
		validate: ({ value }) => normalizeMattermostBaseUrl(value) ? void 0 : "Mattermost base URL must include a valid base URL.",
		normalizeValue: ({ value }) => normalizeMattermostBaseUrl(value) ?? value.trim()
	}],
	disable: (cfg) => ({
		...cfg,
		channels: {
			...cfg.channels,
			mattermost: {
				...cfg.channels?.mattermost,
				enabled: false
			}
		}
	})
};
//#endregion
//#region extensions/mattermost/src/channel.ts
const mattermostSecurityAdapter = createRestrictSendersChannelSecurity({
	channelKey: "mattermost",
	resolveDmPolicy: (account) => account.config.dmPolicy,
	resolveDmAllowFrom: (account) => account.config.allowFrom,
	resolveGroupPolicy: (account) => account.config.groupPolicy,
	surface: "Mattermost channels",
	openScope: "any member",
	groupPolicyPath: "channels.mattermost.groupPolicy",
	groupAllowFromPath: "channels.mattermost.groupAllowFrom",
	policyPathSuffix: "dmPolicy",
	normalizeDmEntry: (raw) => normalizeAllowEntry(raw)
});
function describeMattermostMessageTool({ cfg, accountId }) {
	const enabledAccounts = (accountId ? [resolveMattermostAccount({
		cfg,
		accountId
	})] : listMattermostAccountIds(cfg).map((listedAccountId) => resolveMattermostAccount({
		cfg,
		accountId: listedAccountId
	}))).filter((account) => account.enabled).filter((account) => Boolean(account.botToken?.trim() && account.baseUrl?.trim()));
	const actions = [];
	if (enabledAccounts.length > 0) actions.push("send");
	const baseReactions = (cfg.channels?.mattermost?.actions)?.reactions;
	if (enabledAccounts.some((account) => {
		return (account.config.actions?.reactions ?? baseReactions ?? true) !== false;
	})) actions.push("react");
	return {
		actions,
		capabilities: enabledAccounts.length > 0 ? ["buttons"] : [],
		schema: enabledAccounts.length > 0 ? { properties: { buttons: Type.Optional(createMessageToolButtonsSchema()) } } : null
	};
}
const mattermostMessageActions = {
	describeMessageTool: describeMattermostMessageTool,
	supportsAction: ({ action }) => {
		return action === "send" || action === "react";
	},
	handleAction: async ({ action, params, cfg, accountId }) => {
		if (action === "react") {
			const resolvedAccountId = accountId ?? resolveDefaultMattermostAccountId(cfg);
			const mattermostConfig = cfg.channels?.mattermost;
			if (!(resolveMattermostAccount({
				cfg,
				accountId: resolvedAccountId
			}).config.actions?.reactions ?? mattermostConfig?.actions?.reactions ?? true)) throw new Error("Mattermost reactions are disabled in config");
			const { postId, emojiName, remove } = parseMattermostReactActionParams(params);
			if (remove) {
				const result = await removeMattermostReaction({
					cfg,
					postId,
					emojiName,
					accountId: resolvedAccountId
				});
				if (!result.ok) throw new Error(result.error);
				return {
					content: [{
						type: "text",
						text: `Removed reaction :${emojiName}: from ${postId}`
					}],
					details: {}
				};
			}
			const result = await addMattermostReaction({
				cfg,
				postId,
				emojiName,
				accountId: resolvedAccountId
			});
			if (!result.ok) throw new Error(result.error);
			return {
				content: [{
					type: "text",
					text: `Reacted with :${emojiName}: on ${postId}`
				}],
				details: {}
			};
		}
		if (action !== "send") throw new Error(`Unsupported Mattermost action: ${action}`);
		const to = typeof params.to === "string" ? params.to.trim() : typeof params.target === "string" ? params.target.trim() : "";
		if (!to) throw new Error("Mattermost send requires a target (to).");
		const message = typeof params.message === "string" ? params.message : "";
		const replyToId = readMattermostReplyToId(params);
		const resolvedAccountId = accountId || void 0;
		const mediaUrl = typeof params.media === "string" ? params.media.trim() || void 0 : void 0;
		const result = await sendMessageMattermost(to, message, {
			accountId: resolvedAccountId,
			replyToId,
			buttons: Array.isArray(params.buttons) ? params.buttons : void 0,
			attachmentText: typeof params.attachmentText === "string" ? params.attachmentText : void 0,
			mediaUrl
		});
		return {
			content: [{
				type: "text",
				text: JSON.stringify({
					ok: true,
					channel: "mattermost",
					messageId: result.messageId,
					channelId: result.channelId
				})
			}],
			details: {}
		};
	}
};
const meta = {
	id: "mattermost",
	label: "Mattermost",
	selectionLabel: "Mattermost (plugin)",
	detailLabel: "Mattermost Bot",
	docsPath: "/channels/mattermost",
	docsLabel: "mattermost",
	blurb: "self-hosted Slack-style chat; install the plugin to enable.",
	systemImage: "bubble.left.and.bubble.right",
	order: 65,
	quickstartAllowFrom: true
};
function readTrimmedString(value) {
	if (typeof value !== "string") return;
	return value.trim() || void 0;
}
function parseMattermostReactActionParams(params) {
	const postId = readTrimmedString(params.messageId) ?? readTrimmedString(params.postId);
	if (!postId) throw new Error("Mattermost react requires messageId (post id)");
	const emojiName = readTrimmedString(params.emoji)?.replace(/^:+|:+$/g, "");
	if (!emojiName) throw new Error("Mattermost react requires emoji");
	return {
		postId,
		emojiName,
		remove: params.remove === true
	};
}
function readMattermostReplyToId(params) {
	return readTrimmedString(params.replyToId) ?? readTrimmedString(params.replyTo);
}
function normalizeAllowEntry(entry) {
	return entry.trim().replace(/^(mattermost|user):/i, "").replace(/^@/, "").toLowerCase();
}
function formatAllowEntry(entry) {
	const trimmed = entry.trim();
	if (!trimmed) return "";
	if (trimmed.startsWith("@")) {
		const username = trimmed.slice(1).trim();
		return username ? `@${username.toLowerCase()}` : "";
	}
	return trimmed.replace(/^(mattermost|user):/i, "").toLowerCase();
}
const mattermostConfigAdapter = createScopedChannelConfigAdapter({
	sectionKey: "mattermost",
	listAccountIds: listMattermostAccountIds,
	resolveAccount: adaptScopedAccountAccessor(resolveMattermostAccount),
	defaultAccountId: resolveDefaultMattermostAccountId,
	clearBaseFields: [
		"botToken",
		"baseUrl",
		"name"
	],
	resolveAllowFrom: (account) => account.config.allowFrom,
	formatAllowFrom: (allowFrom) => formatNormalizedAllowFromEntries({
		allowFrom,
		normalizeEntry: formatAllowEntry
	})
});
const mattermostPlugin = createChatChannelPlugin({
	base: {
		id: "mattermost",
		meta: { ...meta },
		setup: mattermostSetupAdapter,
		setupWizard: mattermostSetupWizard,
		capabilities: {
			chatTypes: [
				"direct",
				"channel",
				"group",
				"thread"
			],
			reactions: true,
			threads: true,
			media: true,
			nativeCommands: true
		},
		streaming: { blockStreamingCoalesceDefaults: {
			minChars: 1500,
			idleMs: 1e3
		} },
		reload: { configPrefixes: ["channels.mattermost"] },
		configSchema: MattermostChannelConfigSchema,
		config: {
			...mattermostConfigAdapter,
			isConfigured: (account) => Boolean(account.botToken && account.baseUrl),
			describeAccount: (account) => describeAccountSnapshot({
				account,
				configured: Boolean(account.botToken && account.baseUrl),
				extra: {
					botTokenSource: account.botTokenSource,
					baseUrl: account.baseUrl
				}
			})
		},
		auth: mattermostApprovalAuth,
		doctor: mattermostDoctor,
		groups: { resolveRequireMention: resolveMattermostGroupRequireMention },
		actions: mattermostMessageActions,
		secrets: {
			secretTargetRegistryEntries,
			collectRuntimeConfigAssignments
		},
		directory: createChannelDirectoryAdapter({
			listGroups: async (params) => listMattermostDirectoryGroups(params),
			listGroupsLive: async (params) => listMattermostDirectoryGroups(params),
			listPeers: async (params) => listMattermostDirectoryPeers(params),
			listPeersLive: async (params) => listMattermostDirectoryPeers(params)
		}),
		messaging: {
			defaultMarkdownTableMode: "off",
			normalizeTarget: normalizeMattermostMessagingTarget,
			resolveOutboundSessionRoute: (params) => resolveMattermostOutboundSessionRoute(params),
			targetResolver: {
				looksLikeId: looksLikeMattermostTargetId,
				hint: "<channelId|user:ID|channel:ID>",
				resolveTarget: async ({ cfg, accountId, input }) => {
					const resolved = await resolveMattermostOpaqueTarget({
						input,
						cfg,
						accountId
					});
					if (!resolved) return null;
					return {
						to: resolved.to,
						kind: resolved.kind,
						source: "directory"
					};
				}
			}
		},
		status: createComputedAccountStatusAdapter({
			defaultRuntime: createDefaultChannelRuntimeState(DEFAULT_ACCOUNT_ID, {
				connected: false,
				lastConnectedAt: null,
				lastDisconnect: null
			}),
			buildChannelSummary: ({ snapshot }) => buildPassiveProbedChannelStatusSummary(snapshot, {
				botTokenSource: snapshot.botTokenSource ?? "none",
				connected: snapshot.connected ?? false,
				baseUrl: snapshot.baseUrl ?? null
			}),
			probeAccount: async ({ account, timeoutMs }) => {
				const token = account.botToken?.trim();
				const baseUrl = account.baseUrl?.trim();
				if (!token || !baseUrl) return {
					ok: false,
					error: "bot token or baseUrl missing"
				};
				return await probeMattermost(baseUrl, token, timeoutMs, isPrivateNetworkOptInEnabled(account.config));
			},
			resolveAccountSnapshot: ({ account, runtime }) => ({
				accountId: account.accountId,
				name: account.name,
				enabled: account.enabled,
				configured: Boolean(account.botToken && account.baseUrl),
				extra: {
					botTokenSource: account.botTokenSource,
					baseUrl: account.baseUrl,
					connected: runtime?.connected ?? false,
					lastConnectedAt: runtime?.lastConnectedAt ?? null,
					lastDisconnect: runtime?.lastDisconnect ?? null
				}
			})
		}),
		gateway: {
			resolveGatewayAuthBypassPaths: ({ cfg }) => {
				const base = cfg.channels?.mattermost;
				const callbackPaths = new Set(collectMattermostSlashCallbackPaths(base?.commands).filter((path) => path === "/api/channels/mattermost/command" || path.startsWith("/api/channels/mattermost/")));
				const accounts = base?.accounts ?? {};
				for (const account of Object.values(accounts)) {
					const accountConfig = account && typeof account === "object" && !Array.isArray(account) ? account : void 0;
					for (const path of collectMattermostSlashCallbackPaths(accountConfig?.commands)) if (path === "/api/channels/mattermost/command" || path.startsWith("/api/channels/mattermost/")) callbackPaths.add(path);
				}
				return [...callbackPaths];
			},
			startAccount: async (ctx) => {
				const account = ctx.account;
				const statusSink = createAccountStatusSink({
					accountId: ctx.accountId,
					setStatus: ctx.setStatus
				});
				statusSink({
					baseUrl: account.baseUrl,
					botTokenSource: account.botTokenSource
				});
				ctx.log?.info(`[${account.accountId}] starting channel`);
				return monitorMattermostProvider({
					botToken: account.botToken ?? void 0,
					baseUrl: account.baseUrl ?? void 0,
					accountId: account.accountId,
					config: ctx.cfg,
					runtime: ctx.runtime,
					abortSignal: ctx.abortSignal,
					statusSink
				});
			}
		}
	},
	pairing: { text: {
		idLabel: "mattermostUserId",
		message: "OpenClaw: your access has been approved.",
		normalizeAllowEntry: (entry) => normalizeAllowEntry(entry),
		notify: createLoggedPairingApprovalNotifier(({ id }) => `[mattermost] User ${id} approved for pairing`)
	} },
	threading: { scopedAccountReplyToMode: {
		resolveAccount: (cfg, accountId) => resolveMattermostAccount({
			cfg,
			accountId: accountId ?? resolveDefaultMattermostAccountId(cfg)
		}),
		resolveReplyToMode: (account, chatType) => resolveMattermostReplyToMode(account, chatType === "direct" || chatType === "group" || chatType === "channel" ? chatType : "channel")
	} },
	security: mattermostSecurityAdapter,
	outbound: {
		base: {
			deliveryMode: "direct",
			chunker: chunkTextForOutbound,
			chunkerMode: "markdown",
			textChunkLimit: 4e3,
			resolveTarget: ({ to }) => {
				const trimmed = to?.trim();
				if (!trimmed) return {
					ok: false,
					error: /* @__PURE__ */ new Error("Delivering to Mattermost requires --to <channelId|@username|user:ID|channel:ID>")
				};
				return {
					ok: true,
					to: trimmed
				};
			}
		},
		attachedResults: {
			channel: "mattermost",
			sendText: async ({ cfg, to, text, accountId, replyToId, threadId }) => await sendMessageMattermost(to, text, {
				cfg,
				accountId: accountId ?? void 0,
				replyToId: replyToId ?? (threadId != null ? String(threadId) : void 0)
			}),
			sendMedia: async ({ cfg, to, text, mediaUrl, mediaLocalRoots, accountId, replyToId, threadId }) => await sendMessageMattermost(to, text, {
				cfg,
				accountId: accountId ?? void 0,
				mediaUrl,
				mediaLocalRoots,
				replyToId: replyToId ?? (threadId != null ? String(threadId) : void 0)
			})
		}
	}
});
//#endregion
export { isMattermostSenderAllowed, mattermostPlugin };
