// openclaw/dist/model-runtime-Bilsq5sB.js
//#region src/auto-reply/model-runtime.ts
function formatProviderModelRef(providerRaw, modelRaw) {
	const provider = String(providerRaw ?? "").trim();
	const model = String(modelRaw ?? "").trim();
	if (!provider) return model;
	if (!model) return provider;
	const prefix = `${provider}/`;
	if (model.toLowerCase().startsWith(prefix.toLowerCase())) {
		const normalizedModel = model.slice(prefix.length).trim();
		if (normalizedModel) return `${provider}/${normalizedModel}`;
	}
	return `${provider}/${model}`;
}
function normalizeModelWithinProvider(provider, modelRaw) {
	const model = String(modelRaw ?? "").trim();
	if (!provider || !model) return model;
	const prefix = `${provider}/`;
	if (model.toLowerCase().startsWith(prefix.toLowerCase())) {
		const withoutPrefix = model.slice(prefix.length).trim();
		if (withoutPrefix) return withoutPrefix;
	}
	return model;
}
function normalizeModelRef(rawModel, fallbackProvider, parseEmbeddedProvider = false) {
	const trimmed = String(rawModel ?? "").trim();
	const slashIndex = parseEmbeddedProvider ? trimmed.indexOf("/") : -1;
	if (slashIndex > 0) {
		const provider = trimmed.slice(0, slashIndex).trim();
		const model = trimmed.slice(slashIndex + 1).trim();
		if (provider && model) return {
			provider,
			model,
			label: `${provider}/${model}`
		};
	}
	const provider = String(fallbackProvider ?? "").trim();
	const dedupedModel = normalizeModelWithinProvider(provider, trimmed);
	return {
		provider,
		model: dedupedModel || trimmed,
		label: provider ? formatProviderModelRef(provider, dedupedModel || trimmed) : trimmed
	};
}
function resolveSelectedAndActiveModel(params) {
	const selected = normalizeModelRef(params.selectedModel, params.selectedProvider);
	const runtimeModel = params.sessionEntry?.model?.trim();
	const runtimeProvider = params.sessionEntry?.modelProvider?.trim();
	const active = runtimeModel ? normalizeModelRef(runtimeModel, runtimeProvider || selected.provider, !runtimeProvider) : selected;
	return {
		selected,
		active,
		activeDiffers: active.provider !== selected.provider || active.model !== selected.model
	};
}
//#endregion
export { resolveSelectedAndActiveModel as n, formatProviderModelRef as t };
