// openclaw/dist/onboard-DhIeqkNY.js
import { c as applyProviderConfigWithModelCatalogPreset, t as applyAgentDefaultModelPrimary } from "./provider-onboard-Bz8F6rMa.js";
import { a as buildChutesModelDefinition, i as CHUTES_MODEL_CATALOG, o as discoverChutesModels, r as CHUTES_DEFAULT_MODEL_REF, t as CHUTES_BASE_URL } from "./models-D4ZE1CJv.js";
//#region extensions/chutes/provider-catalog.ts
/**
* Build the Chutes provider with dynamic model discovery.
* Falls back to the static catalog on failure.
* Accepts an optional access token (API key or OAuth access token) for authenticated discovery.
*/
async function buildChutesProvider(accessToken) {
	const models = await discoverChutesModels(accessToken);
	return {
		baseUrl: CHUTES_BASE_URL,
		api: "openai-completions",
		models: models.length > 0 ? models : CHUTES_MODEL_CATALOG.map(buildChutesModelDefinition)
	};
}
//#endregion
//#region extensions/chutes/onboard.ts
/**
* Apply Chutes provider configuration without changing the default model.
* Registers all catalog models and sets provider aliases (chutes-fast, etc.).
*/
function applyChutesProviderConfig(cfg) {
	return applyProviderConfigWithModelCatalogPreset(cfg, {
		providerId: "chutes",
		api: "openai-completions",
		baseUrl: CHUTES_BASE_URL,
		catalogModels: CHUTES_MODEL_CATALOG.map(buildChutesModelDefinition),
		aliases: [
			...CHUTES_MODEL_CATALOG.map((model) => `chutes/${model.id}`),
			{
				modelRef: "chutes-fast",
				alias: "chutes/zai-org/GLM-4.7-FP8"
			},
			{
				modelRef: "chutes-vision",
				alias: "chutes/chutesai/Mistral-Small-3.2-24B-Instruct-2506"
			},
			{
				modelRef: "chutes-pro",
				alias: "chutes/deepseek-ai/DeepSeek-V3.2-TEE"
			}
		]
	});
}
/**
* Apply Chutes provider configuration AND set Chutes as the default model.
*/
function applyChutesConfig(cfg) {
	const next = applyChutesProviderConfig(cfg);
	return {
		...next,
		agents: {
			...next.agents,
			defaults: {
				...next.agents?.defaults,
				model: {
					primary: CHUTES_DEFAULT_MODEL_REF,
					fallbacks: ["chutes/deepseek-ai/DeepSeek-V3.2-TEE", "chutes/Qwen/Qwen3-32B"]
				},
				imageModel: {
					primary: "chutes/chutesai/Mistral-Small-3.2-24B-Instruct-2506",
					fallbacks: ["chutes/chutesai/Mistral-Small-3.1-24B-Instruct-2503"]
				}
			}
		}
	};
}
function applyChutesApiKeyConfig(cfg) {
	return applyAgentDefaultModelPrimary(applyChutesProviderConfig(cfg), CHUTES_DEFAULT_MODEL_REF);
}
//#endregion
export { buildChutesProvider as i, applyChutesConfig as n, applyChutesProviderConfig as r, applyChutesApiKeyConfig as t };
